

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API Reference &mdash; optiMHC 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=01f34227"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Examples" href="examples.html" />
    <link rel="prev" title="Usage" href="usage.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            optiMHC
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-optimhc.core">Core Modules</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.core.Pipeline"><code class="docutils literal notranslate"><span class="pre">Pipeline</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.core.Pipeline.__init__"><code class="docutils literal notranslate"><span class="pre">Pipeline.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.core.Pipeline.read_input"><code class="docutils literal notranslate"><span class="pre">Pipeline.read_input()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.core.Pipeline._generate_features"><code class="docutils literal notranslate"><span class="pre">Pipeline._generate_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.core.Pipeline.rescore"><code class="docutils literal notranslate"><span class="pre">Pipeline.rescore()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.core.Pipeline.save_results"><code class="docutils literal notranslate"><span class="pre">Pipeline.save_results()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.core.Pipeline.visualize_results"><code class="docutils literal notranslate"><span class="pre">Pipeline.visualize_results()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.core.Pipeline._run_single_experiment"><code class="docutils literal notranslate"><span class="pre">Pipeline._run_single_experiment()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.core.Pipeline.run"><code class="docutils literal notranslate"><span class="pre">Pipeline.run()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.core.Pipeline.run_experiments"><code class="docutils literal notranslate"><span class="pre">Pipeline.run_experiments()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-optimhc.core.logging_helper">Logging</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.core.logging_helper.setup_loggers"><code class="docutils literal notranslate"><span class="pre">setup_loggers()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.core.logging_helper.debug_logging"><code class="docutils literal notranslate"><span class="pre">debug_logging()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#cli-interface">CLI Interface</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.cli.parse_cli_config"><code class="docutils literal notranslate"><span class="pre">parse_cli_config()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#psm-container">PSM Container</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.psm_container.PsmContainer"><code class="docutils literal notranslate"><span class="pre">PsmContainer</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.psms"><code class="docutils literal notranslate"><span class="pre">PsmContainer.psms</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.__len__"><code class="docutils literal notranslate"><span class="pre">PsmContainer.__len__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.target_psms"><code class="docutils literal notranslate"><span class="pre">PsmContainer.target_psms</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.decoy_psms"><code class="docutils literal notranslate"><span class="pre">PsmContainer.decoy_psms</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.columns"><code class="docutils literal notranslate"><span class="pre">PsmContainer.columns</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.feature_columns"><code class="docutils literal notranslate"><span class="pre">PsmContainer.feature_columns</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.feature_sources"><code class="docutils literal notranslate"><span class="pre">PsmContainer.feature_sources</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.peptides"><code class="docutils literal notranslate"><span class="pre">PsmContainer.peptides</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.ms_data_files"><code class="docutils literal notranslate"><span class="pre">PsmContainer.ms_data_files</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.scan_ids"><code class="docutils literal notranslate"><span class="pre">PsmContainer.scan_ids</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.charges"><code class="docutils literal notranslate"><span class="pre">PsmContainer.charges</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.metadata"><code class="docutils literal notranslate"><span class="pre">PsmContainer.metadata</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.spectrum_ids"><code class="docutils literal notranslate"><span class="pre">PsmContainer.spectrum_ids</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.identifier_columns"><code class="docutils literal notranslate"><span class="pre">PsmContainer.identifier_columns</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.copy"><code class="docutils literal notranslate"><span class="pre">PsmContainer.copy()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.__repr__"><code class="docutils literal notranslate"><span class="pre">PsmContainer.__repr__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.drop_features"><code class="docutils literal notranslate"><span class="pre">PsmContainer.drop_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.drop_source"><code class="docutils literal notranslate"><span class="pre">PsmContainer.drop_source()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.add_metadata"><code class="docutils literal notranslate"><span class="pre">PsmContainer.add_metadata()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.get_top_hits"><code class="docutils literal notranslate"><span class="pre">PsmContainer.get_top_hits()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.add_features"><code class="docutils literal notranslate"><span class="pre">PsmContainer.add_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.add_features_by_index"><code class="docutils literal notranslate"><span class="pre">PsmContainer.add_features_by_index()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.add_results"><code class="docutils literal notranslate"><span class="pre">PsmContainer.add_results()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.psm_container.PsmContainer.write_pin"><code class="docutils literal notranslate"><span class="pre">PsmContainer.write_pin()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#utilities">Utilities</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.utils.convert_pfm_to_pwm"><code class="docutils literal notranslate"><span class="pre">convert_pfm_to_pwm()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.utils.remove_pre_and_nxt_aa"><code class="docutils literal notranslate"><span class="pre">remove_pre_and_nxt_aa()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.utils.remove_modifications"><code class="docutils literal notranslate"><span class="pre">remove_modifications()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.utils.preprocess_peptide"><code class="docutils literal notranslate"><span class="pre">preprocess_peptide()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.utils.list_all_files_in_directory"><code class="docutils literal notranslate"><span class="pre">list_all_files_in_directory()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.utils.extract_unimod_from_peptidoform"><code class="docutils literal notranslate"><span class="pre">extract_unimod_from_peptidoform()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.utils.convert_to_unimod_format"><code class="docutils literal notranslate"><span class="pre">convert_to_unimod_format()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#parser-modules">Parser Modules</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.parser.read_pin"><code class="docutils literal notranslate"><span class="pre">read_pin()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.parser.read_pepxml"><code class="docutils literal notranslate"><span class="pre">read_pepxml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.parser.extract_mzml_data"><code class="docutils literal notranslate"><span class="pre">extract_mzml_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#parser-submodules">Parser Submodules</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.parser.pin.read_pin"><code class="docutils literal notranslate"><span class="pre">read_pin()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.parser.pin._read_single_pin_as_df"><code class="docutils literal notranslate"><span class="pre">_read_single_pin_as_df()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.parser.pepxml.read_pepxml"><code class="docutils literal notranslate"><span class="pre">read_pepxml()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.parser.pepxml._parse_pepxml"><code class="docutils literal notranslate"><span class="pre">_parse_pepxml()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.parser.pepxml._parse_msms_run"><code class="docutils literal notranslate"><span class="pre">_parse_msms_run()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.parser.pepxml._parse_spectrum"><code class="docutils literal notranslate"><span class="pre">_parse_spectrum()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.parser.pepxml._parse_psm"><code class="docutils literal notranslate"><span class="pre">_parse_psm()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.parser.pepxml._log_features"><code class="docutils literal notranslate"><span class="pre">_log_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.parser.mzml.extract_mzml_data"><code class="docutils literal notranslate"><span class="pre">extract_mzml_data()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#feature-generator">Feature Generator</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#feature-generator-submodules">Feature Generator Submodules</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator"><code class="docutils literal notranslate"><span class="pre">BaseFeatureGenerator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.basic.BasicFeatureGenerator"><code class="docutils literal notranslate"><span class="pre">BasicFeatureGenerator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator"><code class="docutils literal notranslate"><span class="pre">DeepLCFeatureGenerator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator"><code class="docutils literal notranslate"><span class="pre">MHCflurryFeatureGenerator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.netMHCpan._predict_peptide_chunk"><code class="docutils literal notranslate"><span class="pre">_predict_peptide_chunk()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator"><code class="docutils literal notranslate"><span class="pre">NetMHCpanFeatureGenerator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.netMHCIIpan._predict_peptide_chunk_class2"><code class="docutils literal notranslate"><span class="pre">_predict_peptide_chunk_class2()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator"><code class="docutils literal notranslate"><span class="pre">NetMHCIIpanFeatureGenerator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator"><code class="docutils literal notranslate"><span class="pre">OverlappingPeptideFeatureGenerator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.overlapping_peptide.assign_brother_aggregated_feature"><code class="docutils literal notranslate"><span class="pre">assign_brother_aggregated_feature()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator"><code class="docutils literal notranslate"><span class="pre">PWMFeatureGenerator</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator"><code class="docutils literal notranslate"><span class="pre">SpectraSimilarityFeatureGenerator</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#rescore-modules">Rescore Modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#visualization">Visualization</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.visualization.plot_feature_importance"><code class="docutils literal notranslate"><span class="pre">plot_feature_importance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.visualization.visualize_target_decoy_features"><code class="docutils literal notranslate"><span class="pre">visualize_target_decoy_features()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.visualization.visualize_feature_correlation"><code class="docutils literal notranslate"><span class="pre">visualize_feature_correlation()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.visualization.save_or_show_plot"><code class="docutils literal notranslate"><span class="pre">save_or_show_plot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#optimhc.visualization.plot_qvalues"><code class="docutils literal notranslate"><span class="pre">plot_qvalues()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">optiMHC</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">API Reference</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="api-reference">
<h1>API Reference<a class="headerlink" href="#api-reference" title="Link to this heading"></a></h1>
<section id="module-optimhc.core">
<span id="core-modules"></span><h2>Core Modules<a class="headerlink" href="#module-optimhc.core" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="optimhc.core.Pipeline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.core.</span></span><span class="sig-name descname"><span class="pre">Pipeline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">config</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/pipeline.html#Pipeline"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.Pipeline" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Main pipeline class for optiMHC, encapsulating the full data processing workflow.</p>
<p>This class orchestrates input parsing, feature generation, rescoring, result saving, and visualization.
It supports both single-run and experiment modes (multiple feature/model combinations).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>config</strong> (<em>str</em><em>, </em><em>dict</em><em>, or </em><em>Config</em>) – Path to YAML config, dict, or Config object.</p>
</dd>
</dl>
<div class="admonition-examples admonition">
<p class="admonition-title">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">optimhc.core</span> <span class="kn">import</span> <span class="n">Pipeline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipeline</span> <span class="o">=</span> <span class="n">Pipeline</span><span class="p">(</span><span class="n">config</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipeline</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="optimhc.core.Pipeline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">config</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/pipeline.html#Pipeline.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.Pipeline.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the pipeline with a configuration file, dict, or Config object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>config</strong> (<em>str</em><em>, </em><em>dict</em><em>, or </em><em>Config</em>) – Path to YAML config, dict, or Config object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.core.Pipeline.read_input">
<span class="sig-name descname"><span class="pre">read_input</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/pipeline.html#Pipeline.read_input"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.Pipeline.read_input" title="Link to this definition"></a></dt>
<dd><p>Read input PSMs based on configuration.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Object containing loaded PSMs.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer">PsmContainer</a></p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ValueError</strong> – If input type is unsupported.</p></li>
<li><p><strong>Exception</strong> – If file reading fails.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.core.Pipeline._generate_features">
<span class="sig-name descname"><span class="pre">_generate_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/pipeline.html#Pipeline._generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.Pipeline._generate_features" title="Link to this definition"></a></dt>
<dd><p>Generate features for PSMs using the configured feature generators.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>psms</strong> (<a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer"><em>PsmContainer</em></a>) – PSM container object.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PSM container with generated features.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer">PsmContainer</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.core.Pipeline.rescore">
<span class="sig-name descname"><span class="pre">rescore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_fdr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rescoring_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/pipeline.html#Pipeline.rescore"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.Pipeline.rescore" title="Link to this definition"></a></dt>
<dd><p>Perform rescoring on the PSMs using the specified or configured model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psms</strong> (<a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer"><em>PsmContainer</em></a>) – PSM container object.</p></li>
<li><p><strong>model_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Model type (‘XGBoost’, ‘RandomForest’, ‘Percolator’).</p></li>
<li><p><strong>n_jobs</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of parallel jobs.</p></li>
<li><p><strong>test_fdr</strong> (<em>float</em><em>, </em><em>optional</em>) – FDR threshold.</p></li>
<li><p><strong>rescoring_features</strong> (<em>list</em><em>, </em><em>optional</em>) – List of features to use for rescoring.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>results</strong> (<em>mokapot.Results</em>) – Rescoring results.</p></li>
<li><p><strong>models</strong> (<em>list</em>) – Trained models.</p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Rescoring logic is adapted from mokapot (<a class="reference external" href="https://mokapot.readthedocs.io/">https://mokapot.readthedocs.io/</a>)</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.core.Pipeline.save_results">
<span class="sig-name descname"><span class="pre">save_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">results</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">models</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_root</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'optimhc'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/pipeline.html#Pipeline.save_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.Pipeline.save_results" title="Link to this definition"></a></dt>
<dd><p>Save rescoring results, PSM data, and trained models to disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psms</strong> (<a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer"><em>PsmContainer</em></a>) – PSM container object.</p></li>
<li><p><strong>results</strong> (<em>mokapot.Results</em>) – Rescoring results.</p></li>
<li><p><strong>models</strong> (<em>list</em>) – Trained models.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Output directory.</p></li>
<li><p><strong>file_root</strong> (<em>str</em><em>, </em><em>optional</em>) – Root name for output files.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.core.Pipeline.visualize_results">
<span class="sig-name descname"><span class="pre">visualize_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">results</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">models</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/pipeline.html#Pipeline.visualize_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.Pipeline.visualize_results" title="Link to this definition"></a></dt>
<dd><p>Generate and save visualizations for the analysis results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psms</strong> (<a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer"><em>PsmContainer</em></a>) – PSM container object.</p></li>
<li><p><strong>results</strong> (<em>mokapot.Results</em>) – Rescoring results.</p></li>
<li><p><strong>models</strong> (<em>list</em>) – Trained models.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Output directory.</p></li>
<li><p><strong>sources</strong> (<em>list</em><em>, </em><em>optional</em>) – Feature sources to include in visualizations.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.core.Pipeline._run_single_experiment">
<span class="sig-name descname"><span class="pre">_run_single_experiment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exp_config</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exp_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exp_dir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/pipeline.html#Pipeline._run_single_experiment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.Pipeline._run_single_experiment" title="Link to this definition"></a></dt>
<dd><p>Run a single experiment with the specified configuration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psms</strong> (<a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer"><em>PsmContainer</em></a>) – PSM container object.</p></li>
<li><p><strong>exp_config</strong> (<em>dict</em>) – Experiment-specific configuration.</p></li>
<li><p><strong>exp_name</strong> (<em>str</em>) – Name of the experiment.</p></li>
<li><p><strong>exp_dir</strong> (<em>str</em>) – Output directory for the experiment.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if experiment succeeded, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.core.Pipeline.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/pipeline.html#Pipeline.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.Pipeline.run" title="Link to this definition"></a></dt>
<dd><p>Run the complete optiMHC pipeline (single run mode).</p>
<p>This method executes the full workflow: input parsing, feature generation, rescoring, saving, and visualization.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><strong>psms</strong> (<em>PsmContainer</em>) – PSM container object.</p></li>
<li><p><strong>results</strong> (<em>mokapot.Results</em>) – Rescoring results.</p></li>
<li><p><strong>models</strong> (<em>list</em>) – Trained models.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.core.Pipeline.run_experiments">
<span class="sig-name descname"><span class="pre">run_experiments</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/pipeline.html#Pipeline.run_experiments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.Pipeline.run_experiments" title="Link to this definition"></a></dt>
<dd><p>Run experiments with different feature/model combinations using multiprocessing.</p>
<p>Each experiment is executed in its own process for complete resource isolation.
The experiment configurations must be provided in the config under the ‘experiments’ key.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-optimhc.core.logging_helper">
<span id="logging"></span><h2>Logging<a class="headerlink" href="#module-optimhc.core.logging_helper" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="optimhc.core.logging_helper.setup_loggers">
<span class="sig-prename descclassname"><span class="pre">optimhc.core.logging_helper.</span></span><span class="sig-name descname"><span class="pre">setup_loggers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">log_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'INFO'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/logging_helper.html#setup_loggers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.logging_helper.setup_loggers" title="Link to this definition"></a></dt>
<dd><p>Create or update all loggers so that each logger has a StreamHandler and optionally a FileHandler.
This ensures all log messages are displayed in the console and optionally saved to a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>log_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the log file. If None, no file logging is set up.</p></li>
<li><p><strong>log_level</strong> (<em>str</em><em>, </em><em>optional</em>) – Logging level (DEBUG, INFO, WARNING, ERROR). Default is “INFO”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.core.logging_helper.debug_logging">
<span class="sig-prename descclassname"><span class="pre">optimhc.core.logging_helper.</span></span><span class="sig-name descname"><span class="pre">debug_logging</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/core/logging_helper.html#debug_logging"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.core.logging_helper.debug_logging" title="Link to this definition"></a></dt>
<dd><p>Print debugging information for all loggers that start with ‘optimhc’ and
the root logger. This helps verify that logger configurations are set properly.</p>
</dd></dl>

</section>
<section id="cli-interface">
<h2>CLI Interface<a class="headerlink" href="#cli-interface" title="Link to this heading"></a></h2>
<dl class="py function" id="module-optimhc.cli">
<dt class="sig sig-object py" id="optimhc.cli.parse_cli_config">
<span class="sig-prename descclassname"><span class="pre">optimhc.cli.</span></span><span class="sig-name descname"><span class="pre">parse_cli_config</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/cli.html#parse_cli_config"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.cli.parse_cli_config" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="psm-container">
<h2>PSM Container<a class="headerlink" href="#psm-container" title="Link to this heading"></a></h2>
<dl class="py class" id="module-optimhc.psm_container">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.psm_container.</span></span><span class="sig-name descname"><span class="pre">PsmContainer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scan_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spectrum_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ms_data_file_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">peptide_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rescoring_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hit_rank_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">charge_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">retention_time_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A container for managing peptide-spectrum matches (PSMs) in immunopeptidomics rescoring pipelines.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psms</strong> (<em>pd.DataFrame</em>) – DataFrame containing the PSM data.</p></li>
<li><p><strong>label_column</strong> (<em>str</em>) – Column containing the label (True for target, False for decoy).</p></li>
<li><p><strong>scan_column</strong> (<em>str</em>) – Column containing the scan number.</p></li>
<li><p><strong>spectrum_column</strong> (<em>str</em>) – Column containing the spectrum identifier.</p></li>
<li><p><strong>ms_data_file_column</strong> (<em>str</em>) – Column containing the MS data file that the PSM originated from.</p></li>
<li><p><strong>peptide_column</strong> (<em>str</em>) – Column containing the peptide sequence.</p></li>
<li><p><strong>protein_column</strong> (<em>str</em>) – Column containing the protein accessions.</p></li>
<li><p><strong>rescoring_features</strong> (<em>dict</em><em> of </em><em>str to list</em><em> of </em><em>str</em>) – Dictionary of feature columns for rescoring.</p></li>
<li><p><strong>hit_rank_column</strong> (<em>str</em><em>, </em><em>optional</em>) – Column containing the hit rank.</p></li>
<li><p><strong>charge_column</strong> (<em>str</em><em>, </em><em>optional</em>) – Column containing the charge state.</p></li>
<li><p><strong>retention_time_column</strong> (<em>str</em><em>, </em><em>optional</em>) – Column containing the retention time.</p></li>
<li><p><strong>metadata_column</strong> (<em>str</em><em>, </em><em>optional</em>) – Column containing metadata.</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>psms</strong> (<em>pd.DataFrame</em>) – Copy of the DataFrame containing the PSM data.</p></li>
<li><p><strong>target_psms</strong> (<em>pd.DataFrame</em>) – DataFrame containing only target PSMs (label = True).</p></li>
<li><p><strong>decoy_psms</strong> (<em>pd.DataFrame</em>) – DataFrame containing only decoy PSMs (label = False).</p></li>
<li><p><strong>peptides</strong> (<em>list</em><em> of </em><em>str</em>) – List containing all peptides from the PSM data.</p></li>
<li><p><strong>columns</strong> (<em>list</em><em> of </em><em>str</em>) – List of column names in the PSM DataFrame.</p></li>
<li><p><strong>rescoring_features</strong> (<em>dict</em><em> of </em><em>str to list</em><em> of </em><em>str</em>) – Dictionary of rescoring feature columns in the PSM DataFrame.</p></li>
</ul>
</dd>
</dl>
<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.psms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">psms</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">DataFrame</span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.psms" title="Link to this definition"></a></dt>
<dd><p>Get a copy of the PSM DataFrame to prevent external modification.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of the PSM DataFrame.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.__len__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.__len__" title="Link to this definition"></a></dt>
<dd><p>Get the number of PSMs in the container.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Number of PSMs.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.target_psms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">target_psms</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">DataFrame</span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.target_psms" title="Link to this definition"></a></dt>
<dd><p>Get a DataFrame containing only target PSMs.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame with only target PSMs (label = True).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.decoy_psms">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">decoy_psms</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">DataFrame</span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.decoy_psms" title="Link to this definition"></a></dt>
<dd><p>Get a DataFrame containing only decoy PSMs.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame with only decoy PSMs (label = False).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.columns" title="Link to this definition"></a></dt>
<dd><p>Get the column names of the PSM DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of column names.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.feature_columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.feature_columns" title="Link to this definition"></a></dt>
<dd><p>Get a list of all feature columns in the PSM DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of feature column names.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.feature_sources">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_sources</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.feature_sources" title="Link to this definition"></a></dt>
<dd><p>Get a list of all feature sources in the PSM DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of feature source names.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.peptides">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">peptides</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.peptides" title="Link to this definition"></a></dt>
<dd><p>Get the peptide sequences from the PSM data.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of peptide sequences.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.ms_data_files">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ms_data_files</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.ms_data_files" title="Link to this definition"></a></dt>
<dd><p>Get the MS data files from the PSM data.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of MS data file names.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.scan_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">scan_ids</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.scan_ids" title="Link to this definition"></a></dt>
<dd><p>Get the scan numbers from the PSM data.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of scan numbers.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list of int</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.charges">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">charges</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.charges" title="Link to this definition"></a></dt>
<dd><p>Get the charge states from the PSM data.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of charge states.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list of int</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Series</span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.metadata" title="Link to this definition"></a></dt>
<dd><p>Get the metadata from the PSM data.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Series containing metadata for each PSM.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.Series</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.spectrum_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">spectrum_ids</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.spectrum_ids" title="Link to this definition"></a></dt>
<dd><p>Get the spectrum identifiers from the PSM data.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of spectrum identifiers.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.identifier_columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">identifier_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.psm_container.PsmContainer.identifier_columns" title="Link to this definition"></a></dt>
<dd><p>Get the columns that uniquely identify each PSM.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of identifier column names.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.copy" title="Link to this definition"></a></dt>
<dd><p>Return a deep copy of the PsmContainer object.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A deep copy of the current PsmContainer.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer">PsmContainer</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.__repr__" title="Link to this definition"></a></dt>
<dd><p>Return a string representation of the PsmContainer.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>String summary of the PsmContainer.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.drop_features">
<span class="sig-name descname"><span class="pre">drop_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">features</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.drop_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.drop_features" title="Link to this definition"></a></dt>
<dd><p>Drop specified features from the PSM DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>features</strong> (<em>list</em><em> of </em><em>str</em>) – List of feature column names to drop.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If any of the features do not exist in the DataFrame.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.drop_source">
<span class="sig-name descname"><span class="pre">drop_source</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.drop_source"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.drop_source" title="Link to this definition"></a></dt>
<dd><p>Drop all features associated with a specific source from the PSM DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>source</strong> (<em>str</em>) – Name of the source to drop.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the source does not exist in the rescoring features.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.add_metadata">
<span class="sig-name descname"><span class="pre">add_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadata_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">psms_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.add_metadata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.add_metadata" title="Link to this definition"></a></dt>
<dd><p>Merge new metadata into the PSM DataFrame based on specified columns.
Metadata from the specified source is stored as a nested dictionary inside the metadata column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metadata_df</strong> (<em>pd.DataFrame</em>) – DataFrame containing new metadata to add.</p></li>
<li><p><strong>psms_key</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – Column name(s) in the PSM data to merge on.</p></li>
<li><p><strong>metadata_key</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – Column name(s) in the metadata data to merge on.</p></li>
<li><p><strong>source</strong> (<em>str</em>) – Name of the source of the new metadata.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.get_top_hits">
<span class="sig-name descname"><span class="pre">get_top_hits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.get_top_hits"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.get_top_hits" title="Link to this definition"></a></dt>
<dd><p>Get the top n hits based on the hit rank column.
If the hit rank column is not specified, returns the original PSMs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of top hits to return. Default is 1.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new PsmContainer object containing the top n hits.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer">PsmContainer</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.add_features">
<span class="sig-name descname"><span class="pre">add_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">features_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">psms_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.add_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.add_features" title="Link to this definition"></a></dt>
<dd><p>Merge new features into the PSM DataFrame based on specified columns.</p>
<p>This method performs a left join between the PSM data and feature data,
ensuring that all PSMs are preserved while adding new features. It handles
column name conflicts through optional suffixing and maintains feature source
tracking.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>features_df</strong> (<em>pd.DataFrame</em>) – DataFrame containing new features to add.</p></li>
<li><p><strong>psms_key</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – Column name(s) in the PSM data to merge on.</p></li>
<li><p><strong>feature_key</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – Column name(s) in the features data to merge on.</p></li>
<li><p><strong>source</strong> (<em>str</em>) – Name of the source of the new features (e.g., ‘deeplc’, ‘netmhc’).</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix to add to the new columns if there’s a name conflict.
Required when new feature columns have the same names as existing columns.
For example, if adding features from different sources (e.g., ‘score’ from
DeepLC and NetMHC), use suffixes like ‘_deeplc’ or ‘_netmhc’ to distinguish them.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If duplicate columns exist without suffix.
    If merging features changes the number of PSMs.</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The method follows these steps:
1. Validates input and prepares merge keys
2. Checks for column name conflicts
3. Manages feature source: if the source already exists, it will be overwritten
4. Performs left join merge
5. Verifies data integrity</p>
<p>The suffix parameter is used to handle column name conflicts:
- When adding features from different sources that might have the same column names
- When you want to keep both the original and new features with the same name
- When you need to track the source of features in the column names</p>
<p>If suffix is not provided and there are duplicate column names:
- The method will raise a ValueError
- You must either provide a suffix or rename the columns before adding</p>
</div>
<div class="admonition-examples admonition">
<p class="admonition-title">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">container</span> <span class="o">=</span> <span class="n">PsmContainer</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Adding features without suffix (no conflicts)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">features_df1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>    <span class="s1">&#39;scan&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;feature1&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;feature2&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">]</span>
<span class="gp">... </span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">container</span><span class="o">.</span><span class="n">add_features</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">features_df1</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">psms_key</span><span class="o">=</span><span class="s1">&#39;scan&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">feature_key</span><span class="o">=</span><span class="s1">&#39;scan&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">source</span><span class="o">=</span><span class="s1">&#39;source1&#39;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Adding features with suffix (handling conflicts)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">features_df2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>    <span class="s1">&#39;scan&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">],</span>  <span class="c1"># This would conflict with existing &#39;score&#39;</span>
<span class="gp">... </span>    <span class="s1">&#39;feature3&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.7</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">]</span>
<span class="gp">... </span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">container</span><span class="o">.</span><span class="n">add_features</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">features_df2</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">psms_key</span><span class="o">=</span><span class="s1">&#39;scan&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">feature_key</span><span class="o">=</span><span class="s1">&#39;scan&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">source</span><span class="o">=</span><span class="s1">&#39;source2&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">suffix</span><span class="o">=</span><span class="s1">&#39;_new&#39;</span>  <span class="c1"># &#39;score&#39; becomes &#39;score_new&#39;</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.add_features_by_index">
<span class="sig-name descname"><span class="pre">add_features_by_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">features_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.add_features_by_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.add_features_by_index" title="Link to this definition"></a></dt>
<dd><p>Merge new features into the PSM DataFrame based on the DataFrame index.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>features_df</strong> (<em>pd.DataFrame</em>) – DataFrame containing new features to add.</p></li>
<li><p><strong>source</strong> (<em>str</em>) – Name of the source of the new features.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix to add to the new columns if there’s a name conflict.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.add_results">
<span class="sig-name descname"><span class="pre">add_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">results_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">psms_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">result_key</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.add_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.add_results" title="Link to this definition"></a></dt>
<dd><p>Add results of rescore engine to the PSM DataFrame based on specified columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>results_df</strong> (<em>pd.DataFrame</em>) – DataFrame containing new results to add.</p></li>
<li><p><strong>psms_key</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – Column name(s) in the PSM data to merge on.</p></li>
<li><p><strong>result_key</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – Column name(s) in the results data to merge on.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.psm_container.PsmContainer.write_pin">
<span class="sig-name descname"><span class="pre">write_pin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/psm_container.html#PsmContainer.write_pin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.psm_container.PsmContainer.write_pin" title="Link to this definition"></a></dt>
<dd><p>Write the PSM data to a Percolator input (PIN) file.</p>
<dl class="simple">
<dt>Percolator accepts input in a simple tab-delimited format where each row contains features associated with a single PSM:</dt><dd><p>id &lt;tab&gt; label &lt;tab&gt; scannr &lt;tab&gt; feature1 &lt;tab&gt; … &lt;tab&gt; featureN &lt;tab&gt; peptide &lt;tab&gt; proteinId1 &lt;tab&gt; .. &lt;tab&gt; proteinIdM</p>
</dd>
<dt>With header:</dt><dd><p>SpecID &lt;tab&gt; Label &lt;tab&gt; ScanNr &lt;tab&gt; Feature1 &lt;tab&gt; … &lt;tab&gt; FeatureN &lt;tab&gt; Peptide &lt;tab&gt; Proteins</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output_file</strong> (<em>str</em>) – The path to the output PIN file.</p></li>
<li><p><strong>source</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – List of feature sources to include. If None, include all sources.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The DataFrame written to the PIN file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="utilities">
<h2>Utilities<a class="headerlink" href="#utilities" title="Link to this heading"></a></h2>
<dl class="py function" id="module-optimhc.utils">
<dt class="sig sig-object py" id="optimhc.utils.convert_pfm_to_pwm">
<span class="sig-prename descclassname"><span class="pre">optimhc.utils.</span></span><span class="sig-name descname"><span class="pre">convert_pfm_to_pwm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pfm_filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocount</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_freqs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/utils.html#convert_pfm_to_pwm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.utils.convert_pfm_to_pwm" title="Link to this definition"></a></dt>
<dd><p>Convert a Position Frequency Matrix (PFM) file to a Position Weight Matrix (PWM).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pfm_filename</strong> (<em>str</em>) – The file path to the PFM file.</p></li>
<li><p><strong>pseudocount</strong> (<em>float</em><em>, </em><em>optional</em>) – The pseudocount to add to the PFM to avoid zero probabilities. Default is 0.8.</p></li>
<li><p><strong>background_freqs</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary containing the background frequencies for each amino acid.
If None, uses 1/20 for all.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame representation of the PWM.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The conversion process involves:
1. Adding pseudocounts to the PFM
2. Converting to Position Probability Matrix (PPM)
3. Converting to PWM using log2(PPM/background_freqs)</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.utils.remove_pre_and_nxt_aa">
<span class="sig-prename descclassname"><span class="pre">optimhc.utils.</span></span><span class="sig-name descname"><span class="pre">remove_pre_and_nxt_aa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/utils.html#remove_pre_and_nxt_aa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.utils.remove_pre_and_nxt_aa" title="Link to this definition"></a></dt>
<dd><p>Remove the pre and next amino acids from a peptide sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – The peptide sequence with flanking amino acids.
Example: ‘.AANDAGYFNDEMAPIEVKTK.’</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The peptide sequence with flanking amino acids removed.
Example: ‘AANDAGYFNDEMAPIEVKTK’</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function removes any amino acids before the first ‘.’ and after the last ‘.’
in the peptide sequence.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.utils.remove_modifications">
<span class="sig-prename descclassname"><span class="pre">optimhc.utils.</span></span><span class="sig-name descname"><span class="pre">remove_modifications</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_modification</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/utils.html#remove_modifications"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.utils.remove_modifications" title="Link to this definition"></a></dt>
<dd><p>Remove modifications from a peptide sequence, with an option to keep specific modifications.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptide</strong> (<em>str</em>) – The peptide sequence with modifications in brackets.
Example: ‘AANDAGYFNDEM[15.9949]APIEVK[42.0106]TK’</p></li>
<li><p><strong>keep_modification</strong> (<em>str</em><em> or </em><em>list</em><em>, </em><em>optional</em>) – The modification(s) to keep. If provided, only these modifications will be
preserved in the output sequence. Default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The peptide sequence with modifications removed or kept.
Example: ‘AANDAGYFNDEMAPIEVKTK’ (if keep_modification is None)
Example: ‘AANDAGYFNDEM[15.9949]APIEVKTK’ (if keep_modification=[‘15.9949’])</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Modifications are specified in square brackets after the amino acid.
If keep_modification is provided, only those specific modifications will be
preserved in the output sequence.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.utils.preprocess_peptide">
<span class="sig-prename descclassname"><span class="pre">optimhc.utils.</span></span><span class="sig-name descname"><span class="pre">preprocess_peptide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/utils.html#preprocess_peptide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.utils.preprocess_peptide" title="Link to this definition"></a></dt>
<dd><p>Preprocess the peptide sequence by removing flanking regions and modifications.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – Original peptide sequence with possible flanking regions and modifications.
Example: ‘.AANDAGYFNDEM[15.9949]APIEVK[42.0106]TK.’</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Cleaned peptide sequence without flanking regions and modifications.
Example: ‘AANDAGYFNDEMAPIEVKTK’</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function performs two operations in sequence:
1. Removes flanking amino acids using remove_pre_and_nxt_aa
2. Removes all modifications using remove_modifications</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.utils.list_all_files_in_directory">
<span class="sig-prename descclassname"><span class="pre">optimhc.utils.</span></span><span class="sig-name descname"><span class="pre">list_all_files_in_directory</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/utils.html#list_all_files_in_directory"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.utils.list_all_files_in_directory" title="Link to this definition"></a></dt>
<dd><p>Retrieve all files in the specified directory and return a list of file paths.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>directory_path</strong> (<em>str</em>) – The path to the directory to search in.
Example: ‘/path/to/directory’</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of absolute file paths found in the directory and its subdirectories.
Example: [‘/path/to/directory/file1.txt’, ‘/path/to/directory/subdir/file2.txt’]</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of str</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function recursively searches through all subdirectories and returns
absolute paths for all files found.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.utils.extract_unimod_from_peptidoform">
<span class="sig-prename descclassname"><span class="pre">optimhc.utils.</span></span><span class="sig-name descname"><span class="pre">extract_unimod_from_peptidoform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/utils.html#extract_unimod_from_peptidoform"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.utils.extract_unimod_from_peptidoform" title="Link to this definition"></a></dt>
<dd><p>Convert a modified peptide sequence into DeepLC format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptide</strong> (<em>str</em>) – The input peptide sequence with modifications in brackets.
Example: ‘AANDAGYFNDEM[15.9949]APIEVK[42.0106]TK’</p></li>
<li><p><strong>mod_dict</strong> (<em>dict</em>) – Dictionary mapping modification names (in peptide) to corresponding Unimod names.
Example: {‘15.9949’: ‘Oxidation’, ‘42.0106’: ‘Acetyl’}</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>(seq, modifications):</dt><dd><dl class="simple">
<dt>seq<span class="classifier">str</span></dt><dd><p>The unmodified peptide sequence.</p>
</dd>
<dt>modifications<span class="classifier">str</span></dt><dd><p>String of modifications formatted as <cite>position|UnimodName</cite>, separated by pipes <cite>|</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.utils.convert_to_unimod_format">
<span class="sig-prename descclassname"><span class="pre">optimhc.utils.</span></span><span class="sig-name descname"><span class="pre">convert_to_unimod_format</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/utils.html#convert_to_unimod_format"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.utils.convert_to_unimod_format" title="Link to this definition"></a></dt>
<dd><p>Convert a modified peptide sequence into Unimod format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptide</strong> (<em>str</em>) – The input peptide sequence with modifications in brackets.
Example: ‘AANDAGYFNDEM[15.9949]APIEVK[42.0106]TK’</p></li>
<li><p><strong>mod_dict</strong> (<em>dict</em>) – Dictionary mapping modification names (in peptide) to corresponding Unimod names.
Example: {‘15.9949’: ‘UNIMOD:4’, ‘42.0106’: ‘UNIMOD:1’}</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The peptide sequence formatted for Unimod.
Example: ‘AANDAGYFNDEM[UNIMOD:4]APIEVK[UNIMOD:1]TK’</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function replaces the modification names in brackets with their
corresponding Unimod identifiers while preserving the peptide sequence
structure.</p>
</div>
</dd></dl>

</section>
<section id="parser-modules">
<h2>Parser Modules<a class="headerlink" href="#parser-modules" title="Link to this heading"></a></h2>
<dl class="py function" id="module-optimhc.parser">
<dt class="sig sig-object py" id="optimhc.parser.read_pin">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.</span></span><span class="sig-name descname"><span class="pre">read_pin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pin_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">retention_time_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/pin.html#read_pin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.read_pin" title="Link to this definition"></a></dt>
<dd><p>Read PSMs from a Percolator INput (PIN) file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pin_files</strong> (<em>Union</em><em>[</em><em>str</em><em>, </em><em>List</em><em>[</em><em>str</em><em>]</em><em>, </em><em>pd.DataFrame</em><em>]</em>) – The file path to the PIN file, a list of file paths, or a DataFrame
containing PIN data.</p></li>
<li><p><strong>retention_time_column</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – The column containing the retention time. If None, no retention time
will be included.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A PsmContainer object containing the PSM data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer">PsmContainer</a></p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Reads PIN file(s) into a DataFrame
2. Identifies required columns (case-insensitive)
3. Processes scan IDs and hit ranks
4. Converts data types appropriately
5. Creates a PsmContainer with the processed data</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.parser.read_pepxml">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.</span></span><span class="sig-name descname"><span class="pre">read_pepxml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pepxml_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decoy_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DECOY_'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/pepxml.html#read_pepxml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.read_pepxml" title="Link to this definition"></a></dt>
<dd><p>Read PSMs from a list of PepXML files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pepxml_files</strong> (<em>Union</em><em>[</em><em>str</em><em>, </em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – The file path to the PepXML file or a list of file paths.</p></li>
<li><p><strong>decoy_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – The prefix used to indicate a decoy protein in the description lines
of the FASTA file. Default is “<a href="#id5"><span class="problematic" id="id6">DECOY_</span></a>”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A PsmContainer object containing the PSM data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer">PsmContainer</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the PepXML files were generated by Percolator or PeptideProphet.</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Reads and parses PepXML files
2. Calculates mass difference features
3. Processes matched ions and complementary ions
4. Creates charge columns
5. Log-transforms p-values
6. Returns a PsmContainer with the processed data</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.parser.extract_mzml_data">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.</span></span><span class="sig-name descname"><span class="pre">extract_mzml_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mzml_filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scan_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/mzml.html#extract_mzml_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.extract_mzml_data" title="Link to this definition"></a></dt>
<dd><p>Extract scan data from an mzML file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mzml_filename</strong> (<em>str</em>) – The path to the mzML file.</p></li>
<li><p><strong>scan_ids</strong> (<em>list</em><em>[</em><em>int</em><em>] or </em><em>None</em><em>, </em><em>optional</em>) – A list of scan IDs to extract. If None, extracts all scans.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the extracted scan data with columns:
- source: The source file name
- scan: The scan ID
- mz: The m/z values array
- intensity: The intensity values array
- charge: The charge state
- retention_time: The retention time</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Reads the mzML file using pyteomics
2. Extracts scan data including retention time, charge state, m/z values, and intensities
3. Filters scans based on provided scan IDs if specified
4. Returns a DataFrame with the extracted data</p>
</div>
</dd></dl>

<section id="parser-submodules">
<h3>Parser Submodules<a class="headerlink" href="#parser-submodules" title="Link to this heading"></a></h3>
<dl class="py function" id="module-optimhc.parser.pin">
<dt class="sig sig-object py" id="optimhc.parser.pin.read_pin">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.pin.</span></span><span class="sig-name descname"><span class="pre">read_pin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pin_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">retention_time_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/pin.html#read_pin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.pin.read_pin" title="Link to this definition"></a></dt>
<dd><p>Read PSMs from a Percolator INput (PIN) file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pin_files</strong> (<em>Union</em><em>[</em><em>str</em><em>, </em><em>List</em><em>[</em><em>str</em><em>]</em><em>, </em><em>pd.DataFrame</em><em>]</em>) – The file path to the PIN file, a list of file paths, or a DataFrame
containing PIN data.</p></li>
<li><p><strong>retention_time_column</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – The column containing the retention time. If None, no retention time
will be included.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A PsmContainer object containing the PSM data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer">PsmContainer</a></p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Reads PIN file(s) into a DataFrame
2. Identifies required columns (case-insensitive)
3. Processes scan IDs and hit ranks
4. Converts data types appropriately
5. Creates a PsmContainer with the processed data</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.parser.pin._read_single_pin_as_df">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.pin.</span></span><span class="sig-name descname"><span class="pre">_read_single_pin_as_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pin_file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/pin.html#_read_single_pin_as_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.pin._read_single_pin_as_df" title="Link to this definition"></a></dt>
<dd><p>Read a single PIN file into a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>pin_file</strong> (<em>str</em>) – The file path to the PIN file.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the PSM data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Reads the PIN file header
2. Processes the proteins column as a tab-separated list
3. Creates a DataFrame with the processed data</p>
</div>
</dd></dl>

<dl class="py function" id="module-optimhc.parser.pepxml">
<dt class="sig sig-object py" id="optimhc.parser.pepxml.read_pepxml">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.pepxml.</span></span><span class="sig-name descname"><span class="pre">read_pepxml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pepxml_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decoy_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DECOY_'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/pepxml.html#read_pepxml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.pepxml.read_pepxml" title="Link to this definition"></a></dt>
<dd><p>Read PSMs from a list of PepXML files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pepxml_files</strong> (<em>Union</em><em>[</em><em>str</em><em>, </em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – The file path to the PepXML file or a list of file paths.</p></li>
<li><p><strong>decoy_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – The prefix used to indicate a decoy protein in the description lines
of the FASTA file. Default is “<a href="#id7"><span class="problematic" id="id8">DECOY_</span></a>”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A PsmContainer object containing the PSM data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer">PsmContainer</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the PepXML files were generated by Percolator or PeptideProphet.</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Reads and parses PepXML files
2. Calculates mass difference features
3. Processes matched ions and complementary ions
4. Creates charge columns
5. Log-transforms p-values
6. Returns a PsmContainer with the processed data</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.parser.pepxml._parse_pepxml">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.pepxml.</span></span><span class="sig-name descname"><span class="pre">_parse_pepxml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pepxml_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decoy_prefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/pepxml.html#_parse_pepxml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.pepxml._parse_pepxml" title="Link to this definition"></a></dt>
<dd><p>Parse the PSMs of a PepXML into a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pepxml_file</strong> (<em>str</em>) – The PepXML file to parse.</p></li>
<li><p><strong>decoy_prefix</strong> (<em>str</em>) – The prefix used to indicate a decoy protein in the description lines
of the FASTA file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the information about each PSM.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the file is not a PepXML file or is malformed.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.parser.pepxml._parse_msms_run">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.pepxml.</span></span><span class="sig-name descname"><span class="pre">_parse_msms_run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msms_run</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decoy_prefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/pepxml.html#_parse_msms_run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.pepxml._parse_msms_run" title="Link to this definition"></a></dt>
<dd><p>Parse a single MS/MS run.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>msms_run</strong> (<em>tuple</em><em> of </em><em>anything</em><em>, </em><em>lxml.etree.Element</em>) – The second element of the tuple should be the XML element for a single
msms_run. The first is not used, but is necessary for compatibility
with using <code class="code docutils literal notranslate"><span class="pre">map()</span></code>.</p></li>
<li><p><strong>decoy_prefix</strong> (<em>str</em>) – The prefix used to indicate a decoy protein in the description lines
of the FASTA file.</p></li>
</ul>
</dd>
<dt class="field-even">Yields<span class="colon">:</span></dt>
<dd class="field-even"><p><em>dict</em> – A dictionary describing all of the PSMs in a run.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.parser.pepxml._parse_spectrum">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.pepxml.</span></span><span class="sig-name descname"><span class="pre">_parse_spectrum</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spectrum</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_info</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decoy_prefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/pepxml.html#_parse_spectrum"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.pepxml._parse_spectrum" title="Link to this definition"></a></dt>
<dd><p>Parse the PSMs for a single mass spectrum.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spectrum</strong> (<em>lxml.etree.Element</em>) – The XML element for a single spectrum.</p></li>
<li><p><strong>run_info</strong> (<em>dict</em>) – The parsed run data.</p></li>
<li><p><strong>decoy_prefix</strong> (<em>str</em>) – The prefix used to indicate a decoy protein in the description lines
of the FASTA file.</p></li>
</ul>
</dd>
<dt class="field-even">Yields<span class="colon">:</span></dt>
<dd class="field-even"><p><em>dict</em> – A dictionary describing all of the PSMs for a spectrum.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.parser.pepxml._parse_psm">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.pepxml.</span></span><span class="sig-name descname"><span class="pre">_parse_psm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psm_info</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spec_info</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decoy_prefix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/pepxml.html#_parse_psm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.pepxml._parse_psm" title="Link to this definition"></a></dt>
<dd><p>Parse a single PSM.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psm_info</strong> (<em>lxml.etree.Element</em>) – The XML element containing information about the PSM.</p></li>
<li><p><strong>spec_info</strong> (<em>dict</em>) – The parsed spectrum data.</p></li>
<li><p><strong>decoy_prefix</strong> (<em>str</em>) – The prefix used to indicate a decoy protein in the description lines
of the FASTA file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing parsed data about the PSM.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.parser.pepxml._log_features">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.pepxml.</span></span><span class="sig-name descname"><span class="pre">_log_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/pepxml.html#_log_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.pepxml._log_features" title="Link to this definition"></a></dt>
<dd><p>Log-transform columns that are p-values or E-values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>col</strong> (<em>pandas.Series</em>) – A column of the dataset.</p></li>
<li><p><strong>features</strong> (<em>list</em><em> of </em><em>str</em>) – The features of the dataset. Only feature columns will be considered
for transformation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The log-transformed values of the column if the feature was determined
to be a p-value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.Series</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Detects columns written in scientific notation and log them
2. Uses a simple heuristic to find p-value / E-value features
3. Only transforms if values span &gt;4 orders of magnitude
4. Preserves precision for scientific notation values</p>
</div>
</dd></dl>

<dl class="py function" id="module-optimhc.parser.mzml">
<dt class="sig sig-object py" id="optimhc.parser.mzml.extract_mzml_data">
<span class="sig-prename descclassname"><span class="pre">optimhc.parser.mzml.</span></span><span class="sig-name descname"><span class="pre">extract_mzml_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mzml_filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scan_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/parser/mzml.html#extract_mzml_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.parser.mzml.extract_mzml_data" title="Link to this definition"></a></dt>
<dd><p>Extract scan data from an mzML file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mzml_filename</strong> (<em>str</em>) – The path to the mzML file.</p></li>
<li><p><strong>scan_ids</strong> (<em>list</em><em>[</em><em>int</em><em>] or </em><em>None</em><em>, </em><em>optional</em>) – A list of scan IDs to extract. If None, extracts all scans.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the extracted scan data with columns:
- source: The source file name
- scan: The scan ID
- mz: The m/z values array
- intensity: The intensity values array
- charge: The charge state
- retention_time: The retention time</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Reads the mzML file using pyteomics
2. Extracts scan data including retention time, charge state, m/z values, and intensities
3. Filters scans based on provided scan IDs if specified
4. Returns a DataFrame with the extracted data</p>
</div>
</dd></dl>

</section>
</section>
<section id="feature-generator">
<h2>Feature Generator<a class="headerlink" href="#feature-generator" title="Link to this heading"></a></h2>
<section id="feature-generator-submodules">
<span id="module-optimhc.feature_generator"></span><h3>Feature Generator Submodules<a class="headerlink" href="#feature-generator-submodules" title="Link to this heading"></a></h3>
<dl class="py class" id="module-optimhc.feature_generator.base_feature_generator">
<dt class="sig sig-object py" id="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.base_feature_generator.</span></span><span class="sig-name descname"><span class="pre">BaseFeatureGenerator</span></span><a class="reference internal" href="_modules/optimhc/feature_generator/base_feature_generator.html#BaseFeatureGenerator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">ABC</span></code></p>
<p>Abstract base class for all feature generators in the rescoring pipeline.</p>
<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator.feature_columns">
<em class="property"><span class="pre">abstract</span><span class="w"> </span><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator.feature_columns" title="Link to this definition"></a></dt>
<dd><p>Returns a list of feature names generated by the feature generator.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator.id_column">
<em class="property"><span class="pre">abstract</span><span class="w"> </span><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id_column</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator.id_column" title="Link to this definition"></a></dt>
<dd><p>Returns the column or columns used as key or keys to merge features with PSMs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator.generate_features">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">generate_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/base_feature_generator.html#BaseFeatureGenerator.generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator.generate_features" title="Link to this definition"></a></dt>
<dd><p>Generates features.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></span></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-optimhc.feature_generator.basic">
<dt class="sig sig-object py" id="optimhc.feature_generator.basic.BasicFeatureGenerator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.basic.</span></span><span class="sig-name descname"><span class="pre">BasicFeatureGenerator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_pre_nxt_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_modification</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/basic.html#BasicFeatureGenerator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.basic.BasicFeatureGenerator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator" title="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseFeatureGenerator</span></code></a></p>
<p>Feature generator that generates basic features from peptide sequences.</p>
<p>This generator calculates features such as peptide length, proportion of unique amino acids,
Shannon entropy of amino acid distribution, difference between peptide length and average peptide length,
and count of unique amino acids.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of peptide sequences to generate features for.</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to remove the amino acids adjacent to the peptide.
If True, removes them. Default is True.</p></li>
<li><p><strong>remove_modification</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to remove modifications in the peptide sequences.
If True, removes them. Default is True.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The generated features include:
- length_diff_from_avg: Difference between peptide length and average length
- abs_length_diff_from_avg: Absolute difference between peptide length and average length
- unique_aa_count: Number of unique amino acids in the peptide
- unique_aa_proportion: Proportion of unique amino acids in the peptide
- shannon_entropy: Shannon entropy of amino acid distribution</p>
</div>
<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.basic.BasicFeatureGenerator.feature_columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.basic.BasicFeatureGenerator.feature_columns" title="Link to this definition"></a></dt>
<dd><p>Return the list of generated feature column names.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.basic.BasicFeatureGenerator.id_column">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id_column</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.basic.BasicFeatureGenerator.id_column" title="Link to this definition"></a></dt>
<dd><p>Return the list of input columns required for feature generation.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of input column names required for feature generation.
Currently only requires ‘Peptide’ column.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.basic.BasicFeatureGenerator._preprocess_peptide">
<span class="sig-name descname"><span class="pre">_preprocess_peptide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/basic.html#BasicFeatureGenerator._preprocess_peptide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.basic.BasicFeatureGenerator._preprocess_peptide" title="Link to this definition"></a></dt>
<dd><p>Preprocess peptide sequence by removing adjacent amino acids and modifications.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>peptide</strong> (<em>str</em>)</p>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>peptide (str): Original peptide sequence.</p>
</dd>
<dt>Returns:</dt><dd><p>str: Preprocessed peptide sequence.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.basic.BasicFeatureGenerator._shannon_entropy">
<span class="sig-name descname"><span class="pre">_shannon_entropy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/basic.html#BasicFeatureGenerator._shannon_entropy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.basic.BasicFeatureGenerator._shannon_entropy" title="Link to this definition"></a></dt>
<dd><p>Calculate the Shannon entropy of a peptide sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>sequence</strong> (<em>str</em>)</p>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>sequence (str): Peptide sequence.</p>
</dd>
<dt>Returns:</dt><dd><p>float: Shannon entropy value.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.basic.BasicFeatureGenerator.generate_features">
<span class="sig-name descname"><span class="pre">generate_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/basic.html#BasicFeatureGenerator.generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.basic.BasicFeatureGenerator.generate_features" title="Link to this definition"></a></dt>
<dd><p>Generate basic features for the provided peptides.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing peptides and their computed features:
- length_diff_from_avg: Difference from average peptide length
- abs_length_diff_from_avg: Absolute difference from average length
- unique_aa_count: Number of unique amino acids
- unique_aa_proportion: Proportion of unique amino acids
- shannon_entropy: Shannon entropy of amino acid distribution</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If NaN values are found in the generated features.</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>All features are converted to float type before returning.
The method calculates average peptide length across all peptides
and uses it as a reference for length-based features.</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-optimhc.feature_generator.DeepLC">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.DeepLC.</span></span><span class="sig-name descname"><span class="pre">DeepLCFeatureGenerator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calibration_criteria_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_score_is_better</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calibration_set_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">processes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_pre_nxt_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/DeepLC.html#DeepLCFeatureGenerator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator" title="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseFeatureGenerator</span></code></a></p>
<p>Generate DeepLC-based features for rescoring.</p>
<p>This generator uses DeepLC to predict retention times and calculates various
features based on the differences between predicted and observed retention times.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psms</strong> (<a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer"><em>PsmContainer</em></a>) – PSMs to generate features for.</p></li>
<li><p><strong>calibration_criteria_column</strong> (<em>str</em>) – Column name in the PSMs DataFrame to use for DeepLC calibration.</p></li>
<li><p><strong>lower_score_is_better</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether a lower PSM score denotes a better matching PSM. Default is False.</p></li>
<li><p><strong>calibration_set_size</strong> (<em>int</em><em> or </em><em>float</em><em>, </em><em>optional</em>) – Amount of best PSMs to use for DeepLC calibration. If this value is lower
than the number of available PSMs, all PSMs will be used. Default is 0.15.</p></li>
<li><p><strong>processes</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of processes to use in DeepLC. Default is 1.</p></li>
<li><p><strong>model_path</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the DeepLC model. If None, the default model will be used.</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to remove the first and last amino acids from the peptide sequence.
Default is True.</p></li>
<li><p><strong>mod_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary of modifications to be used for DeepLC. If None, no modifications
will be used.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>DeepLC retraining is on by default. Add <code class="docutils literal notranslate"><span class="pre">deeplc_retrain:</span> <span class="pre">False</span></code> as a keyword
argument to disable retraining.</p>
<p>The generated features include:
- observed_retention_time: Original retention time from the data
- predicted_retention_time: DeepLC predicted retention time
- retention_time_diff: Difference between predicted and observed times
- abs_retention_time_diff: Absolute difference between predicted and observed times
- retention_time_ratio: Ratio of min(pred,obs) to max(pred,obs)</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calibration_criteria_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_score_is_better</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calibration_set_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">processes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_pre_nxt_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/DeepLC.html#DeepLCFeatureGenerator.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.__init__" title="Link to this definition"></a></dt>
<dd><p>Generate DeepLC-based features for rescoring.</p>
<p>DeepLC retraining is on by default. Add <code class="docutils literal notranslate"><span class="pre">deeplc_retrain:</span> <span class="pre">False</span></code> as a keyword argument to
disable retraining.</p>
<p>Parameters:
psms: PsmContainer</p>
<blockquote>
<div><p>PSMs to generate features for.</p>
</div></blockquote>
<dl class="simple">
<dt>calibration_criteria_column: str</dt><dd><p>Column name in the PSMs DataFrame to use for DeepLC calibration.</p>
</dd>
<dt>lower_score_is_better</dt><dd><p>Whether a lower PSM score denotes a better matching PSM. Default: False</p>
</dd>
<dt>calibration_set_size: int or float</dt><dd><p>Amount of best PSMs to use for DeepLC calibration. If this value is lower
than the number of available PSMs, all PSMs will be used. (default: 0.15)</p>
</dd>
<dt>processes: {int, None}</dt><dd><p>Number of processes to use in DeepLC. Defaults to 1.</p>
</dd>
<dt>model_path: str</dt><dd><p>Path to the DeepLC model. If None, the default model will be used.</p>
</dd>
<dt>remove_pre_nxt_aa: bool</dt><dd><p>Whether to remove the first and last amino acids from the peptide sequence.
Default: True</p>
</dd>
<dt>mod_dict: dict</dt><dd><p>Dictionary of modifications to be used for DeepLC. If None, no modifications will be used.</p>
</dd>
<dt><a href="#id1"><span class="problematic" id="id2">*</span></a>args: list</dt><dd><p>Additional positional arguments are passed to DeepLC.</p>
</dd>
<dt>kwargs: dict</dt><dd><p>Additional keyword arguments are passed to DeepLC.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psms</strong> (<a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer"><em>PsmContainer</em></a>)</p></li>
<li><p><strong>calibration_criteria_column</strong> (<em>str</em>)</p></li>
<li><p><strong>lower_score_is_better</strong> (<em>bool</em>)</p></li>
<li><p><strong>calibration_set_size</strong> (<em>int</em><em> | </em><em>float</em><em> | </em><em>None</em>)</p></li>
<li><p><strong>processes</strong> (<em>int</em>)</p></li>
<li><p><strong>model_path</strong> (<em>str</em><em> | </em><em>None</em>)</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em>)</p></li>
<li><p><strong>mod_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>str</em><em>] </em><em>| </em><em>None</em>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.feature_columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.feature_columns" title="Link to this definition"></a></dt>
<dd><p>Return the list of generated feature column names.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of feature column names:
- observed_retention_time
- predicted_retention_time
- retention_time_diff
- abs_retention_time_diff
- retention_time_ratio</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.id_column">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id_column</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.id_column" title="Link to this definition"></a></dt>
<dd><p>Return the list of input columns required for the feature generator.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of input columns required for feature generation.
Currently returns an empty list as the required columns are
handled internally by the PsmContainer.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator._get_deeplc_df">
<span class="sig-name descname"><span class="pre">_get_deeplc_df</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/DeepLC.html#DeepLCFeatureGenerator._get_deeplc_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator._get_deeplc_df" title="Link to this definition"></a></dt>
<dd><p>Extract the format required by DeepLC, while retaining necessary original information.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame with the required DeepLC format and original information:
- original_seq: Original peptide sequence
- label: Target/decoy label
- seq: Cleaned peptide sequence
- modifications: Unimod format modifications
- tr: Retention time
- score: Calibration criteria score</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If retention time column is not found in the PSMs DataFrame.</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method prepares the data in the format required by DeepLC,
including cleaning peptide sequences and converting modifications
to Unimod format.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.generate_features">
<span class="sig-name descname"><span class="pre">generate_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/DeepLC.html#DeepLCFeatureGenerator.generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.generate_features" title="Link to this definition"></a></dt>
<dd><p>Generate DeepLC features for the provided PSMs.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the PSMs with added DeepLC features:
- original_seq: Original peptide sequence
- observed_retention_time: Original retention time
- predicted_retention_time: DeepLC predicted retention time
- retention_time_diff: Difference between predicted and observed times
- abs_retention_time_diff: Absolute difference between predicted and observed times
- retention_time_ratio: Ratio of min(pred,obs) to max(pred,obs)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method:
1. Prepares data in DeepLC format
2. Calibrates DeepLC if calibration set is specified
3. Predicts retention times
4. Calculates various retention time-based features
5. Handles missing values by imputing with median values</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator._get_calibration_psms">
<span class="sig-name descname"><span class="pre">_get_calibration_psms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">deeplc_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/DeepLC.html#DeepLCFeatureGenerator._get_calibration_psms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator._get_calibration_psms" title="Link to this definition"></a></dt>
<dd><p>Get the best scoring PSMs for calibration based on the calibration criteria.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>deeplc_df</strong> (<em>pd.DataFrame</em>) – DataFrame containing DeepLC input data.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame of PSMs selected for calibration, containing only target PSMs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – If calibration_set_size is a float not between 0 and 1.</p></li>
<li><p><strong>TypeError</strong> – If calibration_set_size is neither int nor float.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method:
1. Sorts PSMs based on calibration criteria
2. Selects top N PSMs based on calibration_set_size
3. Filters to keep only target PSMs</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.get_full_data">
<span class="sig-name descname"><span class="pre">get_full_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/DeepLC.html#DeepLCFeatureGenerator.get_full_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.get_full_data" title="Link to this definition"></a></dt>
<dd><p>Get the full DeepLC DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the DeepLC input data with all columns:
- original_seq: Original peptide sequence
- label: Target/decoy label
- seq: Cleaned peptide sequence
- modifications: Unimod format modifications
- tr: Retention time
- score: Calibration criteria score
- predicted_retention_time: DeepLC predicted retention time
- retention_time_diff: Difference between predicted and observed times</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.raw_predictions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">raw_predictions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">DataFrame</span></em><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Get the raw predictions DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the raw predictions:
- peptide: Cleaned peptide sequence
- predicted_rt: DeepLC predicted retention time
- observed_rt: Original retention time
- modifications: Unimod format modifications</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>If predictions haven’t been generated yet, this will trigger
feature generation automatically.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.get_raw_predictions">
<span class="sig-name descname"><span class="pre">get_raw_predictions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/DeepLC.html#DeepLCFeatureGenerator.get_raw_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.get_raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Get the raw predictions DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the raw predictions:
- peptide: Cleaned peptide sequence
- predicted_rt: DeepLC predicted retention time
- observed_rt: Original retention time
- modifications: Unimod format modifications</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This is a convenience method that returns the same data as the
raw_predictions property.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.save_raw_predictions">
<span class="sig-name descname"><span class="pre">save_raw_predictions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/DeepLC.html#DeepLCFeatureGenerator.save_raw_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.DeepLC.DeepLCFeatureGenerator.save_raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Save the raw prediction results to a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> (<em>str</em>) – Path to save the file.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Additional parameters passed to pandas.DataFrame.to_csv.
If ‘index’ is not specified, it defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method saves the raw predictions DataFrame to a CSV file.
The DataFrame includes:
- peptide: Cleaned peptide sequence
- predicted_rt: DeepLC predicted retention time
- observed_rt: Original retention time
- modifications: Unimod format modifications</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-optimhc.feature_generator.mhcflurry">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.mhcflurry.</span></span><span class="sig-name descname"><span class="pre">MHCflurryFeatureGenerator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_pre_nxt_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_modification</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/mhcflurry.html#MHCflurryFeatureGenerator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator" title="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseFeatureGenerator</span></code></a></p>
<p>Generate MHCflurry features for peptides based on specified MHC class I alleles.</p>
<p>This generator calculates MHCflurry presentation scores for each peptide against
the provided MHC class I alleles.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of peptide sequences.</p></li>
<li><p><strong>alleles</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of MHC allele names (e.g., [‘HLA-A01:01’, ‘HLA-B07:02’]).</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to include the previous and next amino acids in peptides.
If True, remove them. Default is True.</p></li>
<li><p><strong>remove_modification</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to include modifications in peptides.
If True, remove them. Default is True.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The generated features include:
- mhcflurry_affinity: Binding affinity score
- mhcflurry_processing_score: Processing score
- mhcflurry_presentation_score: Presentation score
- mhcflurry_presentation_percentile: Presentation percentile</p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.MIN_PEPTIDE_LENGTH">
<span class="sig-name descname"><span class="pre">MIN_PEPTIDE_LENGTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">8</span></em><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.MIN_PEPTIDE_LENGTH" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.MAX_PEPTIDE_LENGTH">
<span class="sig-name descname"><span class="pre">MAX_PEPTIDE_LENGTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">15</span></em><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.MAX_PEPTIDE_LENGTH" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.feature_columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.feature_columns" title="Link to this definition"></a></dt>
<dd><p>Return the list of generated feature column names.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of feature column names:
- mhcflurry_affinity
- mhcflurry_processing_score
- mhcflurry_presentation_score
- mhcflurry_presentation_percentile</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.id_column">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id_column</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.id_column" title="Link to this definition"></a></dt>
<dd><p>Return the list of input columns required for the feature generator.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of input column names.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator._preprocess_peptides">
<span class="sig-name descname"><span class="pre">_preprocess_peptides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/mhcflurry.html#MHCflurryFeatureGenerator._preprocess_peptides"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator._preprocess_peptides" title="Link to this definition"></a></dt>
<dd><p>Preprocess peptide sequence by removing flanking amino acids and modifications.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – Original peptide sequence.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Preprocessed peptide sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator._predict">
<span class="sig-name descname"><span class="pre">_predict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/mhcflurry.html#MHCflurryFeatureGenerator._predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator._predict" title="Link to this definition"></a></dt>
<dd><p>Run MHCflurry predictions using the <cite>predict</cite> method and cache the result.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the prediction results, with the best prediction
for each sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method:
1. Preprocesses peptides
2. Filters peptides by length (8-15 amino acids)
3. Runs MHCflurry predictions
4. Merges results with original peptides</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.raw_predictions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">raw_predictions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">DataFrame</span></em><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Return the raw predictions DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the raw predictions:
- peptide: Cleaned peptide sequence
- allele: MHC allele
- affinity: Binding affinity
- processing_score: Processing score
- presentation_score: Presentation score
- presentation_percentile: Presentation percentile</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.get_raw_predictions">
<span class="sig-name descname"><span class="pre">get_raw_predictions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/mhcflurry.html#MHCflurryFeatureGenerator.get_raw_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.get_raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Get the raw prediction results DataFrame from MHCflurry.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Raw prediction results DataFrame containing:
- peptide: Cleaned peptide sequence
- allele: MHC allele
- affinity: Binding affinity
- processing_score: Processing score
- presentation_score: Presentation score
- presentation_percentile: Presentation percentile</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.save_raw_predictions">
<span class="sig-name descname"><span class="pre">save_raw_predictions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/mhcflurry.html#MHCflurryFeatureGenerator.save_raw_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.save_raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Save the raw prediction results to a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> (<em>str</em>) – Path to save the file.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Additional parameters passed to pandas.DataFrame.to_csv.
If ‘index’ is not specified, it defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method saves the raw predictions DataFrame to a CSV file.
The DataFrame includes:
- peptide: Cleaned peptide sequence
- allele: MHC allele
- affinity: Binding affinity
- processing_score: Processing score
- presentation_score: Presentation score
- presentation_percentile: Presentation percentile</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.generate_features">
<span class="sig-name descname"><span class="pre">generate_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/mhcflurry.html#MHCflurryFeatureGenerator.generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.generate_features" title="Link to this definition"></a></dt>
<dd><p>Generate MHCflurry features for the provided peptides and alleles.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the peptides and their predicted MHCflurry features:
- Peptide: Original peptide sequence
- mhcflurry_affinity: Binding affinity
- mhcflurry_processing_score: Processing score
- mhcflurry_presentation_score: Presentation score
- mhcflurry_presentation_percentile: Presentation percentile</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method:
1. Runs MHCflurry predictions
2. Renames columns to include ‘<a href="#id9"><span class="problematic" id="id10">mhcflurry_</span></a>’ prefix
3. Fills missing values with median values
4. Returns the final feature DataFrame</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.get_best_allele">
<span class="sig-name descname"><span class="pre">get_best_allele</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/mhcflurry.html#MHCflurryFeatureGenerator.get_best_allele"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.get_best_allele" title="Link to this definition"></a></dt>
<dd><p>Get the best allele for each peptide.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the best alleles for the peptides:
- Peptide: Original peptide sequence
- mhcflurry_best_allele: Best binding allele</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The best allele is determined by the lowest presentation percentile rank.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.predictions_to_dataframe">
<span class="sig-name descname"><span class="pre">predictions_to_dataframe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/mhcflurry.html#MHCflurryFeatureGenerator.predictions_to_dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.mhcflurry.MHCflurryFeatureGenerator.predictions_to_dataframe" title="Link to this definition"></a></dt>
<dd><p>Convert the predictions to a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the predictions.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If no predictions are available.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function" id="module-optimhc.feature_generator.netMHCpan">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan._predict_peptide_chunk">
<span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.netMHCpan.</span></span><span class="sig-name descname"><span class="pre">_predict_peptide_chunk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides_chunk</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#_predict_peptide_chunk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan._predict_peptide_chunk" title="Link to this definition"></a></dt>
<dd><p>Predict NetMHCpan scores for a chunk of peptides.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides_chunk</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of peptide sequences.</p></li>
<li><p><strong>alleles</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of MHC allele names.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing predictions:
- peptide: Peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.netMHCpan.</span></span><span class="sig-name descname"><span class="pre">NetMHCpanFeatureGenerator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'best'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_pre_nxt_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_modification</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_processes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator" title="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseFeatureGenerator</span></code></a></p>
<p>Generate NetMHCpan features for peptides based on specified MHC class I alleles.</p>
<p>This generator calculates NetMHCpan binding predictions for each peptide against
the provided MHC class I alleles.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of peptide sequences.</p></li>
<li><p><strong>alleles</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of MHC allele names (e.g., [‘HLA-A*02:01’, ‘HLA-B*07:02’]).</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – Mode of feature generation. Options:
- ‘best’: Return only the best allele information for each peptide.
- ‘all’: Return predictions for all alleles with allele-specific suffixes plus best allele info.
Default is ‘best’.</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to include the previous and next amino acids in peptides.
If True, remove them. Default is True.</p></li>
<li><p><strong>remove_modification</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to include modifications in peptides.
If True, remove them. Default is True.</p></li>
<li><p><strong>n_processes</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of processes to use for multiprocessing.
Default is 1 (no multiprocessing).</p></li>
<li><p><strong>show_progress</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to display a progress bar. Default is False.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The generated features include:
- netmhcpan_score: Raw binding score
- netmhcpan_affinity: Binding affinity in nM
- netmhcpan_percentile_rank: Percentile rank of the binding score</p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.MIN_PEPTIDE_LENGTH">
<span class="sig-name descname"><span class="pre">MIN_PEPTIDE_LENGTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">8</span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.MIN_PEPTIDE_LENGTH" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.MAX_PEPTIDE_LENGTH">
<span class="sig-name descname"><span class="pre">MAX_PEPTIDE_LENGTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">30</span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.MAX_PEPTIDE_LENGTH" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.CHUNKSIZE">
<span class="sig-name descname"><span class="pre">CHUNKSIZE</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">250</span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.CHUNKSIZE" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.feature_columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.feature_columns" title="Link to this definition"></a></dt>
<dd><p>Return the list of generated feature column names, determined by the mode.
Only includes numerical features, excluding any string features like allele names.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>List of feature column names:
- For ‘all’ mode: netmhcpan_score_{allele}, netmhcpan_affinity_{allele},</p>
<blockquote>
<div><p>netmhcpan_percentile_rank_{allele} for each allele</p>
</div></blockquote>
<ul class="simple">
<li><p>For both modes: netmhcpan_best_score, netmhcpan_best_affinity,
netmhcpan_best_percentile_rank</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.id_column">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id_column</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.id_column" title="Link to this definition"></a></dt>
<dd><p>Return the list of input columns required for the feature generator.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of input column names.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._preprocess_peptides">
<span class="sig-name descname"><span class="pre">_preprocess_peptides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator._preprocess_peptides"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._preprocess_peptides" title="Link to this definition"></a></dt>
<dd><p>Preprocess the input peptide by removing flanking amino acids,
modifications, and replacing non-standard amino acids.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – The original peptide sequence.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The preprocessed peptide sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._predict_multiprocessing">
<span class="sig-name descname"><span class="pre">_predict_multiprocessing</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides_to_predict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator._predict_multiprocessing"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._predict_multiprocessing" title="Link to this definition"></a></dt>
<dd><p>Run NetMHCpan predictions using multiprocessing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>peptides_to_predict</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of peptides to predict.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing the prediction results:
- peptide: Peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method:
1. Splits peptides into chunks
2. Processes chunks in parallel
3. Combines results into a single DataFrame</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._predict">
<span class="sig-name descname"><span class="pre">_predict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator._predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._predict" title="Link to this definition"></a></dt>
<dd><p>Run NetMHCpan predictions and cache the result.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the prediction results:
- Peptide: Original peptide sequence
- peptide: Cleaned peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method:
1. Preprocesses peptides
2. Filters peptides by length (8-30 amino acids)
3. Runs predictions (with or without multiprocessing)
4. Merges results with original peptides</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.raw_predictions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">raw_predictions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">DataFrame</span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Return the raw prediction results from NetMHCpan.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Raw prediction results DataFrame containing:
- peptide: Cleaned peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.get_raw_predictions">
<span class="sig-name descname"><span class="pre">get_raw_predictions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator.get_raw_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.get_raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Get the raw prediction results DataFrame from NetMHCpan.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Raw prediction results DataFrame containing:
- peptide: Cleaned peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.save_raw_predictions">
<span class="sig-name descname"><span class="pre">save_raw_predictions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator.save_raw_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.save_raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Save the raw prediction results to a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> (<em>str</em>) – Path to save the file.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Additional parameters passed to pandas.DataFrame.to_csv.
If ‘index’ is not specified, it defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method saves the raw predictions DataFrame to a CSV file.
The DataFrame includes:
- peptide: Cleaned peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.generate_features">
<span class="sig-name descname"><span class="pre">generate_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator.generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.generate_features" title="Link to this definition"></a></dt>
<dd><p>Generate the final feature table with NetMHCpan features for each peptide.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>DataFrame containing peptides and their predicted features:
- Peptide: Original peptide sequence
- For ‘all’ mode: netmhcpan_score_{allele}, netmhcpan_affinity_{allele},</p>
<blockquote>
<div><p>netmhcpan_percentile_rank_{allele} for each allele</p>
</div></blockquote>
<ul class="simple">
<li><p>For both modes: netmhcpan_best_score, netmhcpan_best_affinity,
netmhcpan_best_percentile_rank</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The features generated depend on the mode:
- ‘best’: Only the best allele information for each peptide
- ‘all’: All allele predictions plus best allele information</p>
<p>Missing values are handled consistently by filling with median values
for numeric columns.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._generate_all_allele_features">
<span class="sig-name descname"><span class="pre">_generate_all_allele_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">predictions_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator._generate_all_allele_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._generate_all_allele_features" title="Link to this definition"></a></dt>
<dd><p>Generate features for all alleles.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>predictions_df</strong> (<em>pd.DataFrame</em>) – The predictions DataFrame.</p></li>
<li><p><strong>features_df</strong> (<em>pd.DataFrame</em>) – The features DataFrame to update.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Updated features DataFrame with all allele features:
- Peptide: Original peptide sequence
- netmhcpan_score_{allele}: Raw binding score for each allele
- netmhcpan_affinity_{allele}: Binding affinity for each allele
- netmhcpan_percentile_rank_{allele}: Percentile rank for each allele</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._generate_best_allele_features">
<span class="sig-name descname"><span class="pre">_generate_best_allele_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">predictions_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator._generate_best_allele_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._generate_best_allele_features" title="Link to this definition"></a></dt>
<dd><p>Generate features for the best allele.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>predictions_df</strong> (<em>pd.DataFrame</em>) – The predictions DataFrame.</p></li>
<li><p><strong>features_df</strong> (<em>pd.DataFrame</em>) – The features DataFrame to update.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Updated features DataFrame with best allele features:
- Peptide: Original peptide sequence
- netmhcpan_best_allele: Best binding allele
- netmhcpan_best_score: Best binding score
- netmhcpan_best_affinity: Best binding affinity
- netmhcpan_best_percentile_rank: Best percentile rank</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The best allele is determined by the lowest percentile rank.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._fill_missing_values">
<span class="sig-name descname"><span class="pre">_fill_missing_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">features_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator._fill_missing_values"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator._fill_missing_values" title="Link to this definition"></a></dt>
<dd><p>Fill missing values in the features DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>features_df</strong> (<em>pd.DataFrame</em>) – The features DataFrame to update.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Updated features DataFrame with filled missing values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method:
1. Fills best allele string values with ‘Unknown’
2. Fills numeric values with median for all allele features
3. Fills numeric values for best allele features with median</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.predictions_to_dataframe">
<span class="sig-name descname"><span class="pre">predictions_to_dataframe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCpan.html#NetMHCpanFeatureGenerator.predictions_to_dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCpan.NetMHCpanFeatureGenerator.predictions_to_dataframe" title="Link to this definition"></a></dt>
<dd><p>Convert the predictions to a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the predictions.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If no predictions are available.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function" id="module-optimhc.feature_generator.netMHCIIpan">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan._predict_peptide_chunk_class2">
<span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.netMHCIIpan.</span></span><span class="sig-name descname"><span class="pre">_predict_peptide_chunk_class2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides_chunk</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#_predict_peptide_chunk_class2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan._predict_peptide_chunk_class2" title="Link to this definition"></a></dt>
<dd><p>Use NetMHCIIpan43_BA to predict a batch of peptides (MHC Class II).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>peptides_chunk</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>alleles</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>peptides_chunk (List[str]): A batch of peptide sequences to predict.
alleles (List[str]): List of MHC Class II alleles, e.g., [‘DRB1_0101’, ‘DRB1_0102’].</p>
</dd>
<dt>Returns:</dt><dd><p>pd.DataFrame: A DataFrame containing prediction results.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.netMHCIIpan.</span></span><span class="sig-name descname"><span class="pre">NetMHCIIpanFeatureGenerator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'best'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_pre_nxt_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_modification</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_processes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator" title="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseFeatureGenerator</span></code></a></p>
<p>Generate NetMHCIIpan features for given peptides based on specified MHC Class II alleles.</p>
<p>This feature generator uses the NetMHCIIpan43_BA interface to predict MHC Class II binding
for each peptide and returns scores and features based on the specified parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of peptide sequences.</p></li>
<li><p><strong>alleles</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of MHC Class II alleles, e.g., [‘DRB1_0101’, ‘DRB1_0102’].</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – Feature generation mode. Options:
- ‘best’: Return only the best result for each peptide across all alleles.
- ‘all’: Return prediction results for each peptide across all alleles (with allele-specific column suffixes).
Default is ‘best’.</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to remove the amino acids flanking the peptide (e.g., removing X-AA/AA-X forms).
Default is True.</p></li>
<li><p><strong>remove_modification</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to remove modification information from peptides, e.g., (Phospho).
Default is True.</p></li>
<li><p><strong>n_processes</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of processes to use. Default is 1 (no multiprocessing).</p></li>
<li><p><strong>show_progress</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to display a progress bar. Default is False.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The generated features include:
- netmhciipan_score: Raw binding score
- netmhciipan_affinity: Binding affinity in nM
- netmhciipan_percentile_rank: Percentile rank of the binding score</p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.MIN_PEPTIDE_LENGTH">
<span class="sig-name descname"><span class="pre">MIN_PEPTIDE_LENGTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">9</span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.MIN_PEPTIDE_LENGTH" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.MAX_PEPTIDE_LENGTH">
<span class="sig-name descname"><span class="pre">MAX_PEPTIDE_LENGTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">50</span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.MAX_PEPTIDE_LENGTH" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.CHUNKSIZE">
<span class="sig-name descname"><span class="pre">CHUNKSIZE</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">500</span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.CHUNKSIZE" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.feature_columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.feature_columns" title="Link to this definition"></a></dt>
<dd><p>Return the list of generated feature column names, determined by the mode.
Only includes numerical features, excluding any string features like allele names.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>List of feature column names:
- For ‘all’ mode: netmhciipan_score_{allele}, netmhciipan_affinity_{allele},</p>
<blockquote>
<div><p>netmhciipan_percentile_rank_{allele} for each allele</p>
</div></blockquote>
<ul class="simple">
<li><p>For both modes: netmhciipan_best_score, netmhciipan_best_affinity,
netmhciipan_best_percentile_rank</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.id_column">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id_column</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.id_column" title="Link to this definition"></a></dt>
<dd><p>Return the list of input columns required for the feature generator.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of input column names.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._preprocess_peptides">
<span class="sig-name descname"><span class="pre">_preprocess_peptides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator._preprocess_peptides"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._preprocess_peptides" title="Link to this definition"></a></dt>
<dd><p>Preprocess the input peptide by removing flanking amino acids,
modifications, and replacing non-standard amino acids.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – The original peptide sequence.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The preprocessed peptide sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._predict_multiprocessing">
<span class="sig-name descname"><span class="pre">_predict_multiprocessing</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides_to_predict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator._predict_multiprocessing"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._predict_multiprocessing" title="Link to this definition"></a></dt>
<dd><p>Run NetMHCIIpan predictions using multiprocessing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>peptides_to_predict</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of peptides to predict.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing the prediction results:
- peptide: Peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method:
1. Splits peptides into chunks
2. Processes chunks in parallel
3. Combines results into a single DataFrame</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._predict">
<span class="sig-name descname"><span class="pre">_predict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator._predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._predict" title="Link to this definition"></a></dt>
<dd><p>Run NetMHCIIpan predictions and cache the result.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the prediction results:
- Peptide: Original peptide sequence
- peptide: Cleaned peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method:
1. Preprocesses peptides
2. Filters peptides by length (9-30 amino acids)
3. Runs predictions (with or without multiprocessing)
4. Merges results with original peptides</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.generate_features">
<span class="sig-name descname"><span class="pre">generate_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator.generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.generate_features" title="Link to this definition"></a></dt>
<dd><p>Generate the final feature table with NetMHCIIpan features for each peptide.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>DataFrame containing peptides and their predicted features:
- Peptide: Original peptide sequence
- For ‘all’ mode: netmhciipan_score_{allele}, netmhciipan_affinity_{allele},</p>
<blockquote>
<div><p>netmhciipan_percentile_rank_{allele} for each allele</p>
</div></blockquote>
<ul class="simple">
<li><p>For both modes: netmhciipan_best_score, netmhciipan_best_affinity,
netmhciipan_best_percentile_rank</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The features generated depend on the mode:
- ‘best’: Only the best allele information for each peptide
- ‘all’: All allele predictions plus best allele information</p>
<p>Missing values are handled consistently by filling with median values
for numeric columns.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._generate_all_allele_features">
<span class="sig-name descname"><span class="pre">_generate_all_allele_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">predictions_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator._generate_all_allele_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._generate_all_allele_features" title="Link to this definition"></a></dt>
<dd><p>Generate features for all alleles.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>predictions_df</strong> (<em>pd.DataFrame</em>) – The predictions DataFrame.</p></li>
<li><p><strong>features_df</strong> (<em>pd.DataFrame</em>) – The features DataFrame to update.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Updated features DataFrame with all allele features:
- Peptide: Original peptide sequence
- netmhciipan_score_{allele}: Raw binding score for each allele
- netmhciipan_affinity_{allele}: Binding affinity for each allele
- netmhciipan_percentile_rank_{allele}: Percentile rank for each allele</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._generate_best_allele_features">
<span class="sig-name descname"><span class="pre">_generate_best_allele_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">predictions_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator._generate_best_allele_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._generate_best_allele_features" title="Link to this definition"></a></dt>
<dd><p>Generate features for the best allele.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>predictions_df</strong> (<em>pd.DataFrame</em>) – The predictions DataFrame.</p></li>
<li><p><strong>features_df</strong> (<em>pd.DataFrame</em>) – The features DataFrame to update.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Updated features DataFrame with best allele features:
- Peptide: Original peptide sequence
- netmhciipan_best_allele: Best binding allele
- netmhciipan_best_score: Best binding score
- netmhciipan_best_affinity: Best binding affinity
- netmhciipan_best_percentile_rank: Best percentile rank</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The best allele is determined by the lowest percentile rank.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._fill_missing_values">
<span class="sig-name descname"><span class="pre">_fill_missing_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">features_df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator._fill_missing_values"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator._fill_missing_values" title="Link to this definition"></a></dt>
<dd><p>Fill missing values in the features DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>features_df</strong> (<em>pd.DataFrame</em>) – The features DataFrame to update.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Updated features DataFrame with filled missing values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method:
1. Fills best allele string values with ‘Unknown’
2. Fills numeric values with median for all allele features
3. Fills numeric values for best allele features with median</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.predictions_to_dataframe">
<span class="sig-name descname"><span class="pre">predictions_to_dataframe</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator.predictions_to_dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.predictions_to_dataframe" title="Link to this definition"></a></dt>
<dd><p>Convert the predictions to a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the predictions.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If no predictions are available.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.raw_predictions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">raw_predictions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">DataFrame</span></em><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Return the raw prediction results from NetMHCIIpan.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Raw prediction results DataFrame containing:
- peptide: Cleaned peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.get_raw_predictions">
<span class="sig-name descname"><span class="pre">get_raw_predictions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator.get_raw_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.get_raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Get the raw prediction results DataFrame from NetMHCIIpan.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Raw prediction results DataFrame containing:
- peptide: Cleaned peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.save_raw_predictions">
<span class="sig-name descname"><span class="pre">save_raw_predictions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/netMHCIIpan.html#NetMHCIIpanFeatureGenerator.save_raw_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.netMHCIIpan.NetMHCIIpanFeatureGenerator.save_raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Save the raw prediction results to a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> (<em>str</em>) – Path to save the file.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Additional parameters passed to pandas.DataFrame.to_csv.
If ‘index’ is not specified, it defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method saves the raw predictions DataFrame to a CSV file.
The DataFrame includes:
- peptide: Cleaned peptide sequence
- allele: MHC allele
- score: Raw binding score
- affinity: Binding affinity in nM
- percentile_rank: Percentile rank</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-optimhc.feature_generator.overlapping_peptide">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.overlapping_peptide.</span></span><span class="sig-name descname"><span class="pre">OverlappingPeptideFeatureGenerator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overlap_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_entropy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'median'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_pre_nxt_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_modification</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator" title="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseFeatureGenerator</span></code></a></p>
<p>Generates features based on peptide sequence overlaps using the Overlap-Layout-Consensus (OLC) algorithm.</p>
<p>This generator constructs an overlap graph of peptides, removes transitive edges, simplifies the graph to contigs,
and computes features such as the number of overlaps, log-transformed overlap counts, overlap ranks, and contig lengths.
It also filters out peptides with low entropy or outlier lengths before processing.
Additionally, it records detailed information about brother peptides and contigs, accessible via the get_all_data method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>list</em><em> of </em><em>str</em>) – List of peptide sequences.</p></li>
<li><p><strong>min_overlap_length</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum required overlap length for peptides to be considered overlapping. Default is 6.</p></li>
<li><p><strong>min_length</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum peptide length to include in processing. Default is 7.</p></li>
<li><p><strong>max_length</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum peptide length to include in processing. Default is 60.</p></li>
<li><p><strong>min_entropy</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum Shannon entropy for peptides to include in processing. Default is 0.</p></li>
<li><p><strong>fill_missing</strong> (<em>str</em><em>, </em><em>optional</em>) – Method to fill missing values for filtered peptides. Options are ‘median’ or ‘zero’. Default is ‘median’.</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to remove the preceding and following amino acids from peptides. Default is False.</p></li>
<li><p><strong>remove_modification</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to remove modifications from peptides. Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>original_peptides</strong> (<em>list</em><em> of </em><em>str</em>) – Original list of peptide sequences.</p></li>
<li><p><strong>min_overlap_length</strong> (<em>int</em>) – Minimum required overlap length.</p></li>
<li><p><strong>min_length</strong> (<em>int</em>) – Minimum peptide length.</p></li>
<li><p><strong>max_length</strong> (<em>int</em>) – Maximum peptide length.</p></li>
<li><p><strong>min_entropy</strong> (<em>float</em>) – Minimum Shannon entropy.</p></li>
<li><p><strong>fill_missing</strong> (<em>str</em>) – Method to fill missing values.</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em>) – Whether to remove preceding and following amino acids.</p></li>
<li><p><strong>remove_modification</strong> (<em>bool</em>) – Whether to remove modifications.</p></li>
<li><p><strong>filtered_peptides</strong> (<em>list</em><em> of </em><em>str</em>) – List of peptides after filtering.</p></li>
<li><p><strong>filtered_indices</strong> (<em>list</em><em> of </em><em>int</em>) – Indices of filtered peptides.</p></li>
<li><p><strong>peptide_to_index</strong> (<em>dict</em><em> of </em><em>str to int</em>) – Mapping of peptides to their indices.</p></li>
<li><p><strong>overlap_data</strong> (<em>pd.DataFrame</em>) – DataFrame containing overlap data.</p></li>
<li><p><strong>peptide_to_contig</strong> (<em>dict</em><em> of </em><em>str to int</em>) – Mapping of peptides to their contig indices.</p></li>
<li><p><strong>assembled_contigs</strong> (<em>list</em><em> of </em><em>dict</em>) – List of assembled contigs.</p></li>
<li><p><strong>full_data</strong> (<em>pd.DataFrame</em>) – Full data including brother peptides and contig information.</p></li>
<li><p><strong>_overlap_graph</strong> (<em>nx.DiGraph</em>) – Overlap graph.</p></li>
<li><p><strong>_simplified_graph</strong> (<em>nx.DiGraph</em>) – Simplified graph with transitive edges removed.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<dl>
<dt>Key Data Structures:</dt><dd><ol class="arabic">
<li><p>contigs: List[List[str]]
- Represents non-branching paths in the overlap graph
- Each inner list contains peptide sequences that form a continuous chain
- Example: [[‘PEPTIDE1’, ‘PEPTIDE2’], [‘PEPTIDE3’]]</p></li>
<li><p>assembled_contigs: List[Dict]
- Contains the assembled sequences and their constituent peptides
- Each dictionary has two keys:</p>
<blockquote>
<div><p>‘sequence’: The merged/assembled sequence of overlapping peptides
‘peptides’: List of peptides that were used to build this contig</p>
</div></blockquote>
<ul>
<li><dl>
<dt>Example: [</dt><dd><dl class="simple">
<dt>{</dt><dd><p>‘sequence’: ‘LONGPEPTIDESEQUENCE’,
‘peptides’: [‘LONGPEP’, ‘PEPTIDE’, ‘SEQUENCE’]</p>
</dd>
</dl>
<p>},
{</p>
<blockquote>
<div><p>‘sequence’: ‘SINGLEPEPTIDE’,
‘peptides’: [‘SINGLEPEPTIDE’]</p>
</div></blockquote>
<p>}</p>
</dd>
</dl>
</li>
</ul>
<p>]</p>
</li>
<li><p>peptide_to_contig: Dict[str, int]
- Maps each peptide to its contig index in assembled_contigs
- Key: peptide sequence
- Value: index of the contig containing this peptide
- Example: {</p>
<blockquote>
<div><p>‘LONGPEP’: 0,
‘PEPTIDE’: 0,
‘SEQUENCE’: 0,
‘SINGLEPEPTIDE’: 1</p>
</div></blockquote>
<p>}</p>
</li>
<li><p>overlap_graph (_overlap_graph): nx.DiGraph
- Directed graph representing all possible overlaps between peptides
- Nodes: peptide sequences
- Edges: overlaps between peptides
- Edge weights: length of overlap</p></li>
<li><p>simplified_graph (_simplified_graph): nx.DiGraph
- Simplified version of overlap_graph with transitive edges removed
- Used for final contig assembly
- More efficient representation of essential overlaps</p></li>
</ol>
</dd>
</dl>
</div>
<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.id_column">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id_column</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.id_column" title="Link to this definition"></a></dt>
<dd><p>Returns a list of input columns required for the feature generator.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>List[str]: List of input columns.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.feature_columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.feature_columns" title="Link to this definition"></a></dt>
<dd><p>Returns the feature column names.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.overlap_graph">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">overlap_graph</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">DiGraph</span></em><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.overlap_graph" title="Link to this definition"></a></dt>
<dd><p>Returns the overlap graph.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.simplified_graph">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">simplified_graph</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">DiGraph</span></em><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.simplified_graph" title="Link to this definition"></a></dt>
<dd><p>Returns the layout graph.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.contigs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">contigs</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Dict</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.contigs" title="Link to this definition"></a></dt>
<dd><p>Returns the assembled contigs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._shannon_entropy">
<span class="sig-name descname"><span class="pre">_shannon_entropy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._shannon_entropy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._shannon_entropy" title="Link to this definition"></a></dt>
<dd><p>Calculate the Shannon entropy of a peptide sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>sequence</strong> (<em>str</em>)</p>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>sequence (str): Peptide sequence.</p>
</dd>
<dt>Returns:</dt><dd><p>float: Shannon entropy value.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._filter_peptides">
<span class="sig-name descname"><span class="pre">_filter_peptides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._filter_peptides"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._filter_peptides" title="Link to this definition"></a></dt>
<dd><p>Filter out peptides based on length and entropy.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>peptides (List[str]): List of peptide sequences.</p>
</dd>
<dt>Returns:</dt><dd><p>List[str]: Filtered list of peptide sequences.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._construct_prefix_index">
<span class="sig-name descname"><span class="pre">_construct_prefix_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_overlap_length</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._construct_prefix_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._construct_prefix_index" title="Link to this definition"></a></dt>
<dd><p>Construct an index of prefixes for all peptides.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>min_overlap_length</strong> (<em>int</em>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>peptides (List[str]): List of peptide sequences.</p>
</dd>
<dt>Returns:</dt><dd><p>Dict[str, List[int]]: Dictionary mapping prefixes to list of peptide indices.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._build_overlap_graph">
<span class="sig-name descname"><span class="pre">_build_overlap_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix_index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._build_overlap_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._build_overlap_graph" title="Link to this definition"></a></dt>
<dd><p>Build the overlap graph from the list of peptides.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiGraph</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>prefix_index</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>List</em><em>[</em><em>int</em><em>]</em><em>]</em>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>peptides (List[str]): List of peptide sequences.
prefix_index (Dict[str, List[int]]): Index of prefixes.</p>
</dd>
<dt>Returns:</dt><dd><p>nx.DiGraph: Overlap graph.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._remove_transitive_edges">
<span class="sig-name descname"><span class="pre">_remove_transitive_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._remove_transitive_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._remove_transitive_edges" title="Link to this definition"></a></dt>
<dd><p>Remove transitive edges from the overlap graph G.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiGraph</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>G</strong> (<em>DiGraph</em>)</p>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>G (nx.DiGraph): Overlap graph.</p>
</dd>
<dt>Returns:</dt><dd><p>nx.DiGraph: Simplified graph with transitive edges removed.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._simplify_graph_to_contigs">
<span class="sig-name descname"><span class="pre">_simplify_graph_to_contigs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._simplify_graph_to_contigs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._simplify_graph_to_contigs" title="Link to this definition"></a></dt>
<dd><p>Simplify the graph by finding non-branching paths (contigs).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>G</strong> (<em>DiGraph</em>)</p>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>G (nx.DiGraph): Overlap graph.</p>
</dd>
<dt>Returns:</dt><dd><p>List[List[str]]: List of contigs, each contig is a list of peptide sequences.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._assemble_contigs">
<span class="sig-name descname"><span class="pre">_assemble_contigs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contigs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">G</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._assemble_contigs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._assemble_contigs" title="Link to this definition"></a></dt>
<dd><p>Assemble sequences for each contig.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>contigs</strong> (<em>List</em><em>[</em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>)</p></li>
<li><p><strong>G</strong> (<em>DiGraph</em>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>contigs (List[List[str]]): List of contigs.</p>
</dd>
<dt>Returns:</dt><dd><p>List[Dict]: List of assembled contigs with sequence and peptides.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._map_peptides_to_contigs">
<span class="sig-name descname"><span class="pre">_map_peptides_to_contigs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembled_contigs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._map_peptides_to_contigs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._map_peptides_to_contigs" title="Link to this definition"></a></dt>
<dd><p>Map each peptide to its contig.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><p>assembled_contigs (List[Dict]): List of assembled contigs.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>assembled_contigs</strong> (<em>List</em><em>[</em><em>Dict</em><em>]</em>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._remove_redundant_peptides">
<span class="sig-name descname"><span class="pre">_remove_redundant_peptides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._remove_redundant_peptides"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._remove_redundant_peptides" title="Link to this definition"></a></dt>
<dd><p>Remove peptides that are fully contained in other peptides.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>peptides (List[str]): List of peptide sequences.</p>
</dd>
<dt>Returns:</dt><dd><p>accepted_peptides (List[str]): List of accepted peptides not redundant (largest container peptides).
redundant_mapping (Dict[str, str]): Mapping of redundant peptide to its largest container peptide.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._build_full_contig_map">
<span class="sig-name descname"><span class="pre">_build_full_contig_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._build_full_contig_map"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._build_full_contig_map" title="Link to this definition"></a></dt>
<dd><p>Build a mapping from contig index to list of peptides (both accepted and redundant).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>peptides (List[str]): Original list of peptides.</p>
</dd>
<dt>Returns:</dt><dd><p>Dict[int, List[str]]: Mapping from contig index to list of peptides.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._calculate_overlap_contig_features">
<span class="sig-name descname"><span class="pre">_calculate_overlap_contig_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._calculate_overlap_contig_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._calculate_overlap_contig_features" title="Link to this definition"></a></dt>
<dd><p>Calculate overlap and contig related features for peptides.</p>
<p>This method computes the overlap graph, simplifies it to contigs, assembles contigs,
and generates features for each peptide. It also handles redundant peptides by mapping them
to their container peptides.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>peptides (List[str]): List of peptide sequences.</p>
</dd>
<dt>Returns:</dt><dd><p>pd.DataFrame: DataFrame containing the computed features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._integrate_overlap_features">
<span class="sig-name descname"><span class="pre">_integrate_overlap_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator._integrate_overlap_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator._integrate_overlap_features" title="Link to this definition"></a></dt>
<dd><p>Compute the features for each peptide, ensuring output aligns with input.</p>
<p>This method preprocesses and filters peptides, calculates overlap and contig features,
maps them back to the original peptides, and fills missing values.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></span></p>
</dd>
</dl>
<dl class="simple">
<dt>Returns:</dt><dd><p>pd.DataFrame: DataFrame containing features for each peptide.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.generate_features">
<span class="sig-name descname"><span class="pre">generate_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator.generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.generate_features" title="Link to this definition"></a></dt>
<dd><p>Generates features for peptide overlaps, including the count of overlapping peptides, contig length,
and log-transformed counts and ranks.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></span></p>
</dd>
</dl>
<dl class="simple">
<dt>Returns:</dt><dd><p>pd.DataFrame: DataFrame containing the features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.get_full_data">
<span class="sig-name descname"><span class="pre">get_full_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#OverlappingPeptideFeatureGenerator.get_full_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.OverlappingPeptideFeatureGenerator.get_full_data" title="Link to this definition"></a></dt>
<dd><p>Returns the full data including brother peptides and contig information for each peptide.
In the output, the lists of contig peptides and brother peptides include redundant peptides,
so that their counts match the corresponding peptide and contig_member_count.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></span></p>
</dd>
</dl>
<dl class="simple">
<dt>Returns:</dt><dd><p>pd.DataFrame: DataFrame containing peptides and their brother peptides and contigs.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.feature_generator.overlapping_peptide.assign_brother_aggregated_feature">
<span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.overlapping_peptide.</span></span><span class="sig-name descname"><span class="pre">assign_brother_aggregated_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_columns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlapping_source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'OverlappingGroupFeatures'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/overlapping_peptide.html#assign_brother_aggregated_feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.overlapping_peptide.assign_brother_aggregated_feature" title="Link to this definition"></a></dt>
<dd><p>Assign aggregated features based on brother peptides to the PSMs.</p>
<p>For PSMs with the same ContigSequence (brother peptides), compute the mean of specified features
and assign these aggregated features back to each PSM in the group. Additionally, compute
the sum as mean * (contig_member_count + 1). If a PSM does not have a ContigSequence (no brothers),
its new features will be set to the original values.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>psms</strong> (<a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer"><em>PsmContainer</em></a>)</p></li>
<li><p><strong>feature_columns</strong> (<em>str</em><em> | </em><em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>overlapping_source</strong> (<em>str</em>)</p></li>
<li><p><strong>source_name</strong> (<em>str</em>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>psms (PsmContainer): PSM container containing the peptides and features.
feature_columns (Union[str, List[str]]): Name of the feature column(s) to aggregate.
overlapping_source (str): Source name of the overlapping peptide features.
source_name (str): Name of the new feature source.</p>
</dd>
<dt>Returns:</dt><dd><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py class" id="module-optimhc.feature_generator.PWM">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.PWM.</span></span><span class="sig-name descname"><span class="pre">PWMFeatureGenerator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anchors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mhc_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'I'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pwm_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_pre_nxt_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_modification</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator" title="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseFeatureGenerator</span></code></a></p>
<p>Generates PWM (Position Weight Matrix) features for peptides based on specified MHC alleles.</p>
<p>This generator calculates PWM scores for each peptide against the provided MHC class I or II allele PWMs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>list</em><em> of </em><em>str</em>) – Series of peptide sequences.</p></li>
<li><p><strong>alleles</strong> (<em>list</em><em> of </em><em>str</em>) – List of MHC allele names (e.g., [‘HLA-A01:01’, ‘HLA-B07:02’]).</p></li>
<li><p><strong>anchors</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of anchor positions to consider for MHC class I. Default is 2.</p></li>
<li><p><strong>mhc_class</strong> (<em>str</em><em>, </em><em>optional</em>) – MHC class, either ‘I’ or ‘II’. Default is ‘I’.</p></li>
<li><p><strong>pwm_path</strong> (<em>str</em><em> or </em><em>os.PathLike</em><em>, </em><em>optional</em>) – Custom path to PWM files. Defaults to ‘../../data/PWMs’.</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to include the previous and next amino acids in peptides.
If True, remove them. Default is False.</p></li>
<li><p><strong>remove_modification</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to include modifications in peptides.
If True, remove them. Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>peptides</strong> (<em>pd.Series</em>) – Series of peptide sequences.</p></li>
<li><p><strong>alleles</strong> (<em>list</em><em> of </em><em>str</em>) – List of MHC allele names.</p></li>
<li><p><strong>mhc_class</strong> (<em>str</em>) – MHC class (‘I’ or ‘II’).</p></li>
<li><p><strong>pwm_path</strong> (<em>str</em><em> or </em><em>os.PathLike</em>) – Path to PWM files.</p></li>
<li><p><strong>pwms</strong> (<em>dict</em>) – Dictionary of PWMs for each allele and mer length.</p></li>
<li><p><strong>anchors</strong> (<em>int</em>) – Number of anchor positions for MHC class I.</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em>) – Whether to remove pre/post neighboring amino acids.</p></li>
<li><p><strong>remove_modification</strong> (<em>bool</em>) – Whether to remove modifications.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<dl class="simple">
<dt>For MHC class I:</dt><dd><ul class="simple">
<li><p>Generates ‘PWM_Score_{allele}’ and optionally ‘Anchor_Score_{allele}’ columns.</p></li>
</ul>
</dd>
<dt>For MHC class II:</dt><dd><ul class="simple">
<li><p>Generates ‘PWM_Score_{allele}’ (core 9-mer),</p></li>
<li><p>‘N_Flank_PWM_Score_{allele}’,</p></li>
<li><p>‘C_Flank_PWM_Score_{allele}’ columns.</p></li>
</ul>
</dd>
</dl>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator.CLASS_II_CORE_LENGTH">
<span class="sig-name descname"><span class="pre">CLASS_II_CORE_LENGTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">9</span></em><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator.CLASS_II_CORE_LENGTH" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator.DEFAULT_PWM_PATH">
<span class="sig-name descname"><span class="pre">DEFAULT_PWM_PATH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'/data/48/zshang/optiMHC/optimhc/feature_generator/../PWMs'</span></em><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator.DEFAULT_PWM_PATH" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anchors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mhc_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'I'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pwm_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_pre_nxt_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_modification</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator.__init__" title="Link to this definition"></a></dt>
<dd><p>Initializes the PWMFeatureGenerator.</p>
<dl>
<dt>Parameters:</dt><dd><p>peptides (List[str]): Series of peptide sequences.
alleles (List[str]): List of MHC allele names (e.g., [‘HLA-A01:01’, ‘HLA-B07:02’]).
mhc_class (str): MHC class, either ‘I’ or ‘II’. Default is ‘I’.
pwm_path (Optional[Union[str, os.PathLike]]): Custom path to PWM files. Defaults to ‘../../data/PWMs’.
remove_pre_nxt_aa (bool): Whether to include the previous and next amino acids in peptides.</p>
<blockquote>
<div><p>If True, remove them. Default is False.</p>
</div></blockquote>
<dl class="simple">
<dt>remove_modification (bool): Whether to include modifications in peptides.</dt><dd><p>If True, remove them. Default is True.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>alleles</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>anchors</strong> (<em>int</em>)</p></li>
<li><p><strong>mhc_class</strong> (<em>str</em>)</p></li>
<li><p><strong>pwm_path</strong> (<em>str</em><em> | </em><em>PathLike</em><em> | </em><em>None</em>)</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em>)</p></li>
<li><p><strong>remove_modification</strong> (<em>bool</em>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator.id_column">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id_column</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator.id_column" title="Link to this definition"></a></dt>
<dd><p>Get a list of input columns required for the feature generator.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of column names required for feature generation.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list of str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator._extract_trailing_numbers">
<span class="sig-name descname"><span class="pre">_extract_trailing_numbers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator._extract_trailing_numbers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator._extract_trailing_numbers" title="Link to this definition"></a></dt>
<dd><p>Extract the trailing numbers from a string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>text</strong> (<em>str</em>) – Input string to extract numbers from.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The trailing numbers if found, None otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str or None</p>
</dd>
</dl>
<div class="admonition-examples admonition">
<p class="admonition-title">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">_extract_trailing_numbers</span><span class="p">(</span><span class="s1">&#39;ABC123&#39;</span><span class="p">)</span>
<span class="go">&#39;123&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_extract_trailing_numbers</span><span class="p">(</span><span class="s1">&#39;ABC&#39;</span><span class="p">)</span>
<span class="go">None</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_extract_trailing_numbers</span><span class="p">(</span><span class="s1">&#39;L123&#39;</span><span class="p">)</span>
<span class="go">&#39;123&#39;</span>
</pre></div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator._default_allele_pwm_files">
<span class="sig-name descname"><span class="pre">_default_allele_pwm_files</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator._default_allele_pwm_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator._default_allele_pwm_files" title="Link to this definition"></a></dt>
<dd><p>Construct default PWM file paths for each allele based on MHC class.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping alleles to their PWM files for each mer length.
Format: {allele: {mer_length: file_path}}</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict of str to dict of int to str</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<dl class="simple">
<dt>For MHC class I:</dt><dd><ul class="simple">
<li><p>Allele directory format: HLA-A01:01 -&gt; HLA-A01_01</p></li>
</ul>
</dd>
<dt>For MHC class II:</dt><dd><ul class="simple">
<li><p>Allele directory format: DRB10101 -&gt; DRB1_0101</p></li>
<li><p>Fixed core length of 9</p></li>
</ul>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator._most_conserved_postions">
<span class="sig-name descname"><span class="pre">_most_conserved_postions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pwm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator._most_conserved_postions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator._most_conserved_postions" title="Link to this definition"></a></dt>
<dd><p>Find the n most conserved positions in the PWM.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pwm</strong> (<em>pd.DataFrame</em>) – Position Weight Matrix to analyze.</p></li>
<li><p><strong>n</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of positions to return. Default is 2.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Indices of the n most conserved positions.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of int</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If n exceeds the PWM length.</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>In our study, we only use the anchor score for class I MHC.
Conservation is measured using Shannon entropy.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator._load_pwms">
<span class="sig-name descname"><span class="pre">_load_pwms</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator._load_pwms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator._load_pwms" title="Link to this definition"></a></dt>
<dd><p>Load PWMs for each allele from the constructed file paths.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary of PWMs for each allele and mer length.
Format: {allele: {mer_length: pwm_dataframe}}</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict of str to dict of int to pd.DataFrame</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If a PWM file is not found.</p></li>
<li><p><strong>Exception</strong> – If there is an error loading a PWM file.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>PWM files are expected to be space-delimited text files with amino acids as row indices
and positions as column indices.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator._cal_PWM_score_I">
<span class="sig-name descname"><span class="pre">_cal_PWM_score_I</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allele</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator._cal_PWM_score_I"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator._cal_PWM_score_I" title="Link to this definition"></a></dt>
<dd><p>Calculate PWM scores for MHC class I.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptide</strong> (<em>str</em>) – The peptide sequence to score.</p></li>
<li><p><strong>allele</strong> (<em>str</em>) – The MHC allele to score against.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PWM score for the peptide against the allele’s PWM, or None if out of range.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float or None</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>If peptide length is out of range, returns None.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator._cal_PWM_score_II">
<span class="sig-name descname"><span class="pre">_cal_PWM_score_II</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allele</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator._cal_PWM_score_II"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator._cal_PWM_score_II" title="Link to this definition"></a></dt>
<dd><p>Calculate PWM scores for MHC class II using a sliding 9-mer window.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptide</strong> (<em>str</em>) – The peptide sequence to score.</p></li>
<li><p><strong>allele</strong> (<em>str</em>) – The MHC allele to score against.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A tuple containing:
- core_score : float or None</p>
<blockquote>
<div><p>Score for the best 9-mer core.</p>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>n_flank_score<span class="classifier">float or None</span></dt><dd><p>Score for the N-terminal flanking region.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>c_flank_score<span class="classifier">float or None</span></dt><dd><p>Score for the C-terminal flanking region.</p>
</dd>
</dl>
</li>
</ul>
<p>Returns (None, None, None) if peptide has length &lt; 9.</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple of (float or None, float or None, float or None)</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The method:
1. Slides over all possible 9-mer windows to find the highest core PWM score.
2. Once the best core is found, extracts up to 3 AA on each flank (N-flank and C-flank).
3. If the flank has fewer than 3 residues, pads with ‘X’.
4. Scores each flank with n_flank_pwm and c_flank_pwm.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator._cal_PWM_score">
<span class="sig-name descname"><span class="pre">_cal_PWM_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allele</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator._cal_PWM_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator._cal_PWM_score" title="Link to this definition"></a></dt>
<dd><p>Calculates PWM scores for a single peptide across all applicable mer lengths for a given allele.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>peptide</strong> (<em>str</em>)</p></li>
<li><p><strong>allele</strong> (<em>str</em>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>For MHC class I:</dt><dd><p>Returns a single float (or pd.NA).</p>
</dd>
<dt>For MHC class II:</dt><dd><p>Returns a tuple of (core_score, n_flank_score, c_flank_score) or (pd.NA, pd.NA, pd.NA).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator._cal_anchor_score">
<span class="sig-name descname"><span class="pre">_cal_anchor_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allele</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anchor_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator._cal_anchor_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator._cal_anchor_score" title="Link to this definition"></a></dt>
<dd><p>Calculate anchor score for a single peptide across all applicable mer lengths for a given allele.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptide</strong> (<em>str</em>) – The peptide sequence to score.</p></li>
<li><p><strong>allele</strong> (<em>str</em>) – The MHC allele to score against.</p></li>
<li><p><strong>anchor_dict</strong> (<em>dict</em><em> of </em><em>int to list</em><em> of </em><em>int</em>) – Dictionary containing the most conserved positions for each mer length.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Anchor score for the peptide against the allele’s PWM, or None if out of range.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float or None</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Only implemented for MHC class I. For MHC class II, returns None.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator.set_pwms">
<span class="sig-name descname"><span class="pre">set_pwms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pwms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator.set_pwms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator.set_pwms" title="Link to this definition"></a></dt>
<dd><p>Set PWMs directly, allowing for custom PWMs to be provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>pwms</strong> (<em>dict</em><em> of </em><em>str to dict</em><em> of </em><em>int to pd.DataFrame</em>) – Dictionary of PWMs for each allele and mer length.
Format: {allele: {mer_length: pwm_dataframe}}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator.generate_features">
<span class="sig-name descname"><span class="pre">generate_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/PWM.html#PWMFeatureGenerator.generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator.generate_features" title="Link to this definition"></a></dt>
<dd><p>Generate PWM features for all peptides across specified alleles.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>DataFrame containing generated features:
For MHC class I:</p>
<blockquote>
<div><ul class="simple">
<li><p>’PWM_Score_{allele}’ and optionally ‘Anchor_Score_{allele}’ columns.</p></li>
</ul>
</div></blockquote>
<dl class="simple">
<dt>For MHC class II:</dt><dd><ul class="simple">
<li><p>’PWM_Score_{allele}’ (core 9-mer),</p></li>
<li><p>’N_Flank_PWM_Score_{allele}’,</p></li>
<li><p>’C_Flank_PWM_Score_{allele}’ columns.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Missing values are imputed with the median value for each feature.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.PWM.PWMFeatureGenerator.feature_columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.PWM.PWMFeatureGenerator.feature_columns" title="Link to this definition"></a></dt>
<dd><p>Returns a list of feature names generated by the feature generator.</p>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-optimhc.feature_generator.spectra_similarity">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">optimhc.feature_generator.spectra_similarity.</span></span><span class="sig-name descname"><span class="pre">SpectraSimilarityFeatureGenerator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spectrum_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">charges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scan_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mz_file_paths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collision_energies</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">instruments</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fragmentation_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_pre_nxt_aa</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">url</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'koina.wilhelmlab.org:443'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">36</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolerance_ppm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator" title="optimhc.feature_generator.base_feature_generator.BaseFeatureGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseFeatureGenerator</span></code></a></p>
<p>Feature generator for calculating similarity between experimental and predicted spectra.</p>
<p>This class works through the following steps:
1. Extract experimental spectral data from mzML files
2. Use Koina for theoretical spectra prediction
3. Align experimental and predicted spectra
4. Calculate similarity metrics as features</p>
<dl class="simple">
<dt>Parameters:</dt><dd><p>peptides (List[str]): List of peptide sequences
charges (List[int]): List of charge states
scan_ids (List[int]): List of scan IDs
mz_file_paths (List[str]): List of mzML file paths
model_type (str): Prediction model type, either “HCD” or “CID”
collision_energies (List[float]): List of collision energies, required when model_type is “HCD”
remove_pre_nxt_aa (bool): Whether to remove preceding and next amino acids, default is True
remove_modification (bool): Whether to remove modifications, default is True
url (str): Koina server URL, default is “koina.wilhelmlab.org:443”
top_n (int): Number of top peaks to use for alignment, default is 12
tolerance_ppm (float): Mass tolerance for alignment in ppm, default is 20</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spectrum_ids</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>charges</strong> (<em>List</em><em>[</em><em>int</em><em>]</em>)</p></li>
<li><p><strong>scan_ids</strong> (<em>List</em><em>[</em><em>int</em><em>]</em>)</p></li>
<li><p><strong>mz_file_paths</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>model_type</strong> (<em>str</em>)</p></li>
<li><p><strong>collision_energies</strong> (<em>List</em><em>[</em><em>float</em><em>]</em>)</p></li>
<li><p><strong>instruments</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>fragmentation_types</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>)</p></li>
<li><p><strong>remove_pre_nxt_aa</strong> (<em>bool</em>)</p></li>
<li><p><strong>mod_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>str</em><em>] </em><em>| </em><em>None</em>)</p></li>
<li><p><strong>url</strong> (<em>str</em>)</p></li>
<li><p><strong>top_n</strong> (<em>int</em>)</p></li>
<li><p><strong>tolerance_ppm</strong> (<em>float</em>)</p></li>
</ul>
</dd>
</dl>
<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.id_column">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">id_column</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.id_column" title="Link to this definition"></a></dt>
<dd><p>Returns a list of input columns required for the feature generator.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.feature_columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">feature_columns</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.feature_columns" title="Link to this definition"></a></dt>
<dd><p>Returns a list of feature columns generated by the feature generator.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.input_df">
<span class="sig-name descname"><span class="pre">input_df</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator.input_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.input_df" title="Link to this definition"></a></dt>
<dd><p>Return the generated features as a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></span></p>
</dd>
</dl>
<dl class="simple">
<dt>Returns:</dt><dd><p>pd.DataFrame: DataFrame containing the generated features</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._preprocess_peptide">
<span class="sig-name descname"><span class="pre">_preprocess_peptide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">peptide</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._preprocess_peptide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._preprocess_peptide" title="Link to this definition"></a></dt>
<dd><p>Preprocess peptide sequence.</p>
<p>As Prosit does not support non-standard amino acid ‘U’, we replace it with ‘C’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – Original peptide sequence.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Processed peptide sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This is nonsense when it comes to spectral prediction, but we need to keep it
for compatibility with Koina. In the future, this should be prohibited at the input level.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._extract_experimental_spectra">
<span class="sig-name descname"><span class="pre">_extract_experimental_spectra</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._extract_experimental_spectra"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._extract_experimental_spectra" title="Link to this definition"></a></dt>
<dd><p>Extract experimental spectral data from mzML files.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing experimental spectral data.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The method groups scan IDs by file path for efficiency and extracts
spectral data from each mzML file. The resulting DataFrame contains
m/z values, intensities, and associated metadata for each spectrum.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._predict_theoretical_spectra">
<span class="sig-name descname"><span class="pre">_predict_theoretical_spectra</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">processed_peptides</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">charges</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._predict_theoretical_spectra"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._predict_theoretical_spectra" title="Link to this definition"></a></dt>
<dd><p>Use Koina to predict theoretical spectra.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>processed_peptides</strong> (<em>list</em><em> of </em><em>str</em>) – List of preprocessed peptide sequences.</p></li>
<li><p><strong>charges</strong> (<em>list</em><em> of </em><em>int</em>) – List of charge states.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame containing predicted spectral data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>Exception</strong> – If there is an error during Koina prediction.</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The method uses Koina to predict theoretical spectra for each peptide.
For AlphaPeptDeep_ms2_generic model, inputs are split into batches
grouped by peptide length for prediction.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.raw_predictions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">raw_predictions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">DataFrame</span></em><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Returns the raw prediction results from Koina.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>pd.DataFrame: Raw prediction results DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.get_raw_predictions">
<span class="sig-name descname"><span class="pre">get_raw_predictions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator.get_raw_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.get_raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Get the raw prediction results DataFrame from Koina.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></span></p>
</dd>
</dl>
<dl class="simple">
<dt>Returns:</dt><dd><p>pd.DataFrame: Raw prediction results DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.save_raw_predictions">
<span class="sig-name descname"><span class="pre">save_raw_predictions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator.save_raw_predictions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.save_raw_predictions" title="Link to this definition"></a></dt>
<dd><p>Save the raw prediction results to a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>file_path</strong> (<em>str</em>)</p>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>file_path (str): Path to save the file
<a href="#id3"><span class="problematic" id="id4">**</span></a>kwargs: Other parameters passed to pandas.DataFrame.to_csv</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._sort_spectrum_by_mz">
<span class="sig-name descname"><span class="pre">_sort_spectrum_by_mz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intensity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._sort_spectrum_by_mz"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._sort_spectrum_by_mz" title="Link to this definition"></a></dt>
<dd><p>Sort spectrum (m/z, intensity, and optionally annotation) by m/z values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mz</strong> (<em>list</em><em> of </em><em>float</em>) – m/z values.</p></li>
<li><p><strong>intensity</strong> (<em>list</em><em> of </em><em>float</em>) – Intensity values.</p></li>
<li><p><strong>annotation</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Fragment annotations.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sorted m/z, intensity, and annotation arrays.
If annotation is None, the third element will be None.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple of (np.ndarray, np.ndarray, np.ndarray or None)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._align_spectra_all_peaks">
<span class="sig-name descname"><span class="pre">_align_spectra_all_peaks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_mz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exp_intensity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_mz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_intensity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_ppm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._align_spectra_all_peaks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._align_spectra_all_peaks" title="Link to this definition"></a></dt>
<dd><p>Align experimental and predicted spectra.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>exp_mz</strong> (<em>List</em><em>[</em><em>float</em><em>]</em>)</p></li>
<li><p><strong>exp_intensity</strong> (<em>List</em><em>[</em><em>float</em><em>]</em>)</p></li>
<li><p><strong>pred_mz</strong> (<em>List</em><em>[</em><em>float</em><em>]</em>)</p></li>
<li><p><strong>pred_intensity</strong> (<em>List</em><em>[</em><em>float</em><em>]</em>)</p></li>
<li><p><strong>pred_annotation</strong> (<em>List</em><em>[</em><em>str</em><em>] </em><em>| </em><em>None</em>)</p></li>
<li><p><strong>use_ppm</strong> (<em>bool</em>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>exp_mz (List[float]): Experimental m/z values
exp_intensity (List[float]): Experimental intensity values
pred_mz (List[float]): Predicted m/z values
pred_intensity (List[float]): Predicted intensity values
pred_annotation (Optional[List[str]]): Predicted fragment annotations
use_ppm (bool): Whether to use ppm tolerance or Da tolerance</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>Tuple[np.ndarray, np.ndarray, List[Tuple], Dict]:</dt><dd><ul class="simple">
<li><p>Aligned experimental intensity vector</p></li>
<li><p>Predicted intensity vector</p></li>
<li><p>Matching index pairs</p></li>
<li><p>Additional info including original sorted arrays</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._get_top_peaks_vectors">
<span class="sig-name descname"><span class="pre">_get_top_peaks_vectors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aligned_exp_intensity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aligned_pred_intensity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matched_indices</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._get_top_peaks_vectors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._get_top_peaks_vectors" title="Link to this definition"></a></dt>
<dd><p>Extract top N peaks based on predicted intensity for similarity calculation</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>aligned_exp_intensity</strong> (<em>ndarray</em>)</p></li>
<li><p><strong>aligned_pred_intensity</strong> (<em>ndarray</em>)</p></li>
<li><p><strong>matched_indices</strong> (<em>List</em><em>[</em><em>Tuple</em><em>]</em>)</p></li>
<li><p><strong>top_n</strong> (<em>int</em>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>aligned_exp_intensity (np.ndarray): Aligned experimental intensity vector
aligned_pred_intensity (np.ndarray): Aligned predicted intensity vector
matched_indices (List[Tuple]): Matching index pairs (pred_idx, exp_idx)
top_n (int): Number of top peaks to extract</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>Tuple[np.ndarray, np.ndarray, List[Tuple]]:</dt><dd><ul class="simple">
<li><p>Top N experimental intensity vector</p></li>
<li><p>Top N predicted intensity vector</p></li>
<li><p>Top N matching index pairs</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._align_spectra">
<span class="sig-name descname"><span class="pre">_align_spectra</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_mz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exp_intensity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_mz</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_intensity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._align_spectra"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._align_spectra" title="Link to this definition"></a></dt>
<dd><p>Align experimental and predicted spectra.</p>
<p>This is a wrapper around _align_spectra_all_peaks and _get_top_peaks_vectors
to maintain backward compatibility while using the improved algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_mz</strong> (<em>list</em><em> of </em><em>float</em>) – Experimental m/z values.</p></li>
<li><p><strong>exp_intensity</strong> (<em>list</em><em> of </em><em>float</em>) – Experimental intensity values.</p></li>
<li><p><strong>pred_mz</strong> (<em>list</em><em> of </em><em>float</em>) – Predicted m/z values.</p></li>
<li><p><strong>pred_intensity</strong> (<em>list</em><em> of </em><em>float</em>) – Predicted intensity values.</p></li>
<li><p><strong>pred_annotation</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Predicted fragment annotations.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>Aligned experimental intensity vector</p></li>
<li><p>Predicted intensity vector</p></li>
<li><p>Matching index pairs (for top N peaks)</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple of (np.ndarray, np.ndarray, list of tuple)</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The method first aligns all peaks using _align_spectra_all_peaks, then
extracts the top N peaks using _get_top_peaks_vectors for compatibility
with existing code.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._normalize_vector_l2">
<span class="sig-name descname"><span class="pre">_normalize_vector_l2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._normalize_vector_l2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._normalize_vector_l2" title="Link to this definition"></a></dt>
<dd><p>Normalize a vector using L2 normalization (unit vector).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vector</strong> (<em>np.ndarray</em>) – Input vector to normalize.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>L2-normalized vector.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>If the input vector has zero norm, the original vector is returned unchanged.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._normalize_vector_sum">
<span class="sig-name descname"><span class="pre">_normalize_vector_sum</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._normalize_vector_sum"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._normalize_vector_sum" title="Link to this definition"></a></dt>
<dd><p>Perform sum normalization (probability normalization) on a vector.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vector</strong> (<em>np.ndarray</em>) – Input vector.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sum normalized vector (sum to 1).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>If the input vector has zero sum, the original vector is returned unchanged.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_spectral_angle_similarity">
<span class="sig-name descname"><span class="pre">_calculate_spectral_angle_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._calculate_spectral_angle_similarity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_spectral_angle_similarity" title="Link to this definition"></a></dt>
<dd><p>Calculate the spectral angle between experimental and predicted vectors.</p>
<p>Normalize the angle to [0, 1] where 1 is the best similarity.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_vector</strong> (<em>np.ndarray</em>) – Experimental intensity vector.</p></li>
<li><p><strong>pred_vector</strong> (<em>np.ndarray</em>) – Predicted intensity vector.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Spectral angle similarity (0-1, higher is better).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The spectral angle is calculated as: SA = 1 - (2 * angle / π),
where angle is the angle between the normalized vectors in radians.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_cosine_similarity">
<span class="sig-name descname"><span class="pre">_calculate_cosine_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._calculate_cosine_similarity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_cosine_similarity" title="Link to this definition"></a></dt>
<dd><p>Calculate the cosine similarity between experimental and predicted vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_vector</strong> (<em>np.ndarray</em>) – Experimental intensity vector.</p></li>
<li><p><strong>pred_vector</strong> (<em>np.ndarray</em>) – Predicted intensity vector.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Cosine similarity (0-1, higher is better).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The cosine similarity is calculated as 1 - cosine_distance.
If either vector has zero sum, returns 0.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_spearman_correlation">
<span class="sig-name descname"><span class="pre">_calculate_spearman_correlation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._calculate_spearman_correlation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_spearman_correlation" title="Link to this definition"></a></dt>
<dd><p>Calculate Spearman correlation coefficient between experimental and predicted vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_vector</strong> (<em>np.ndarray</em>) – Experimental intensity vector.</p></li>
<li><p><strong>pred_vector</strong> (<em>np.ndarray</em>) – Predicted intensity vector.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Spearman correlation coefficient (-1 to 1, higher is better).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>If either vector has no variation (std = 0), returns 0.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_pearson_correlation">
<span class="sig-name descname"><span class="pre">_calculate_pearson_correlation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._calculate_pearson_correlation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_pearson_correlation" title="Link to this definition"></a></dt>
<dd><p>Calculate Pearson correlation coefficient between experimental and predicted vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_vector</strong> (<em>np.ndarray</em>) – Experimental intensity vector.</p></li>
<li><p><strong>pred_vector</strong> (<em>np.ndarray</em>) – Predicted intensity vector.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pearson correlation coefficient (-1 to 1, higher is better).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>If either vector has no variation (std = 0), returns 0.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_mean_squared_error">
<span class="sig-name descname"><span class="pre">_calculate_mean_squared_error</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._calculate_mean_squared_error"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_mean_squared_error" title="Link to this definition"></a></dt>
<dd><p>Calculate mean squared error between experimental and predicted vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_vector</strong> (<em>np.ndarray</em>) – Experimental intensity vector.</p></li>
<li><p><strong>pred_vector</strong> (<em>np.ndarray</em>) – Predicted intensity vector.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Mean squared error (lower is better).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The vectors are normalized using L2 normalization before calculating
the mean squared error for fair comparison.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_entropy">
<span class="sig-name descname"><span class="pre">_calculate_entropy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._calculate_entropy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_entropy" title="Link to this definition"></a></dt>
<dd><p>Calculate Shannon entropy of a vector that has already been sum-1 normalized.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vector</strong> (<em>np.ndarray</em>) – Input vector, which has already been sum-1 normalized.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Shannon entropy.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Only non-zero probabilities are considered for entropy calculation.
If all probabilities are zero, returns 0.0.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_unweighted_entropy_similarity">
<span class="sig-name descname"><span class="pre">_calculate_unweighted_entropy_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._calculate_unweighted_entropy_similarity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_unweighted_entropy_similarity" title="Link to this definition"></a></dt>
<dd><p>Calculate unweighted spectral entropy between experimental and predicted vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_vector</strong> (<em>np.ndarray</em>) – Experimental intensity vector.</p></li>
<li><p><strong>pred_vector</strong> (<em>np.ndarray</em>) – Predicted intensity vector.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Spectral entropy similarity.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Based on the method described in <a class="reference external" href="https://www.nature.com/articles/s41592-021-01331-z">https://www.nature.com/articles/s41592-021-01331-z</a>.
The spectral entropy is calculated using the formula:
1 - (2*S_PM - S_P - S_M)/ln(4), where S_PM is the entropy of the mixed
distribution, and S_P and S_M are the entropies of the individual distributions.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_predicted_counts">
<span class="sig-name descname"><span class="pre">_calculate_predicted_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._calculate_predicted_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_predicted_counts" title="Link to this definition"></a></dt>
<dd><p>Calculate counts of predicted peaks seen/not seen in experimental spectrum.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_vector</strong> (<em>np.ndarray</em>) – Experimental intensity vector.</p></li>
<li><p><strong>pred_vector</strong> (<em>np.ndarray</em>) – Predicted intensity vector.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>predicted_seen_nonzero: Number of predicted peaks that are also present in experimental spectrum</p></li>
<li><p>predicted_not_seen: Number of predicted peaks that are not present in experimental spectrum</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple of (int, int)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_bray_curtis_similarity">
<span class="sig-name descname"><span class="pre">_calculate_bray_curtis_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._calculate_bray_curtis_similarity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_bray_curtis_similarity" title="Link to this definition"></a></dt>
<dd><p>Calculate Bray-Curtis similarity between experimental and predicted vectors</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>exp_vector</strong> (<em>ndarray</em>)</p></li>
<li><p><strong>pred_vector</strong> (<em>ndarray</em>)</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Parameters:</dt><dd><p>exp_vector (np.ndarray): Experimental intensity vector
pred_vector (np.ndarray): Predicted intensity vector</p>
</dd>
<dt>Returns:</dt><dd><p>float: Bray-Curtis similarity (0-1, higher is better)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_similarity_features">
<span class="sig-name descname"><span class="pre">_calculate_similarity_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._calculate_similarity_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_similarity_features" title="Link to this definition"></a></dt>
<dd><p>Calculate all similarity features between experimental and predicted vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_vector</strong> (<em>np.ndarray</em>) – Experimental intensity vector.</p></li>
<li><p><strong>pred_vector</strong> (<em>np.ndarray</em>) – Predicted intensity vector.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of similarity features, including:
- spectral_angle_similarity: Spectral angle similarity (0-1)
- cosine_similarity: Cosine similarity (0-1)
- pearson_correlation: Pearson correlation (-1 to 1)
- spearman_correlation: Spearman correlation (-1 to 1)
- mean_squared_error: Mean squared error
- unweighted_entropy_similarity: Spectral entropy similarity
- predicted_seen_nonzero: Number of predicted peaks seen in experimental spectrum
- predicted_not_seen: Number of predicted peaks not seen in experimental spectrum
- bray_curtis_similarity: Bray-Curtis similarity (0-1)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict of str to float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._generate_features">
<span class="sig-name descname"><span class="pre">_generate_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator._generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._generate_features" title="Link to this definition"></a></dt>
<dd><p>Generate spectral similarity features</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></span></p>
</dd>
</dl>
<dl class="simple">
<dt>Returns:</dt><dd><p>pd.DataFrame: DataFrame containing generated features</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.generate_features">
<span class="sig-name descname"><span class="pre">generate_features</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator.generate_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.generate_features" title="Link to this definition"></a></dt>
<dd><p>Public interface for generating spectral similarity features.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame containing the generated features.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method is a wrapper around _generate_features that ensures
the results are cached and only computed once.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.get_full_data">
<span class="sig-name descname"><span class="pre">get_full_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/feature_generator/spectra_similarity.html#SpectraSimilarityFeatureGenerator.get_full_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.get_full_data" title="Link to this definition"></a></dt>
<dd><p>Return the full DataFrame with all columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Full DataFrame with all columns.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This method returns the complete DataFrame including all intermediate
results and raw data used in feature generation.</p>
</div>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="rescore-modules">
<h2>Rescore Modules<a class="headerlink" href="#rescore-modules" title="Link to this heading"></a></h2>
</section>
<section id="visualization">
<span id="module-optimhc.rescore"></span><h2>Visualization<a class="headerlink" href="#visualization" title="Link to this heading"></a></h2>
<dl class="py function" id="module-optimhc.visualization">
<dt class="sig sig-object py" id="optimhc.visualization.plot_feature_importance">
<span class="sig-prename descclassname"><span class="pre">optimhc.visualization.</span></span><span class="sig-name descname"><span class="pre">plot_feature_importance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">models</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rescoring_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">error</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/visualization/plot_features.html#plot_feature_importance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.visualization.plot_feature_importance" title="Link to this definition"></a></dt>
<dd><p>Unified function to plot average feature importance across multiple models.</p>
<dl class="simple">
<dt>This function supports:</dt><dd><ul class="simple">
<li><p>Linear models (e.g., Linear SVR) which provide an ‘estimator’ attribute with a ‘<a href="#id11"><span class="problematic" id="id12">coef_</span></a>’.
The absolute value of the coefficients is used for importance, and hatch patterns are applied
to differentiate between positive and negative coefficients.</p></li>
<li><p>XGBoost models which provide a ‘<a href="#id13"><span class="problematic" id="id14">feature_importances_</span></a>’ attribute. Since these values are
always positive, no hatch patterns are applied.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>models</strong> (<em>list</em>) – A list of model objects.
For linear models, each model should have an ‘estimator’ with ‘<a href="#id15"><span class="problematic" id="id16">coef_</span></a>’.
For XGBoost models, each model should have a ‘<a href="#id17"><span class="problematic" id="id18">feature_importances_</span></a>’ attribute.</p></li>
<li><p><strong>rescoring_features</strong> (<em>dict</em>) – A dictionary where keys are sources and values are lists of features.</p></li>
<li><p><strong>save_path</strong> (<em>str</em><em>, </em><em>optional</em>) – If provided, saves the plot to the specified path.</p></li>
<li><p><strong>sort</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, sorts the features by their importance in descending order.
Default is False.</p></li>
<li><p><strong>error</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, adds error bars to the plot. Default is False.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Additional plotting parameters such as ‘figsize’ and ‘dpi’.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The function automatically detects the model type based on the presence of the corresponding attribute.
For linear models, it uses hatch patterns to differentiate between positive and negative coefficients.
For XGBoost models, it uses solid bars since the importances are always positive.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.visualization.visualize_target_decoy_features">
<span class="sig-prename descclassname"><span class="pre">optimhc.visualization.</span></span><span class="sig-name descname"><span class="pre">visualize_target_decoy_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/visualization/plot_tdc_distribution.html#visualize_target_decoy_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.visualization.visualize_target_decoy_features" title="Link to this definition"></a></dt>
<dd><p>Visualize the distribution of features in a DataFrame using kernel density estimation plots.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psms</strong> (<a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer"><em>PsmContainer</em></a>) – A PsmContainer object containing the features to visualize.</p></li>
<li><p><strong>num_cols</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of columns in the plot grid. Default is 5.</p></li>
<li><p><strong>save_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The file path to save the plot. If not provided, the plot is displayed.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Additional plotting parameters such as <cite>figsize</cite> and <cite>dpi</cite>, etc.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Extracts rescoring features from the PsmContainer
2. Filters out features with only one unique value
3. Creates a grid of plots showing the distribution of each feature
4. Separates target and decoy PSMs in each plot
5. Uses kernel density estimation to show the distribution shape</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.visualization.visualize_feature_correlation">
<span class="sig-prename descclassname"><span class="pre">optimhc.visualization.</span></span><span class="sig-name descname"><span class="pre">visualize_feature_correlation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/visualization/plot_features.html#visualize_feature_correlation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.visualization.visualize_feature_correlation" title="Link to this definition"></a></dt>
<dd><p>Visualize the correlation between features in a DataFrame using a heatmap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>psms</strong> (<a class="reference internal" href="#optimhc.psm_container.PsmContainer" title="optimhc.psm_container.PsmContainer"><em>PsmContainer</em></a>) – A PsmContainer object containing the features to visualize.</p></li>
<li><p><strong>save_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The file path to save the plot. If not provided, the plot is displayed.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Additional plotting parameters such as <cite>figsize</cite> and <cite>dpi</cite>, etc.</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Extracts all rescoring features from the PsmContainer
2. Calculates the correlation matrix between features
3. Creates a heatmap visualization of the correlations
4. Uses a coolwarm colormap to show positive and negative correlations</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.visualization.save_or_show_plot">
<span class="sig-prename descclassname"><span class="pre">optimhc.visualization.</span></span><span class="sig-name descname"><span class="pre">save_or_show_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logger</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tight_layout</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/visualization/save_or_show_plot.html#save_or_show_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.visualization.save_or_show_plot" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="optimhc.visualization.plot_qvalues">
<span class="sig-prename descclassname"><span class="pre">optimhc.visualization.</span></span><span class="sig-name descname"><span class="pre">plot_qvalues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">results</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">10)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/optimhc/visualization/plot_roc.html#plot_qvalues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#optimhc.visualization.plot_qvalues" title="Link to this definition"></a></dt>
<dd><p>Plot q-values for the given results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>results</strong> (<em>object</em><em> or </em><em>list</em>) – A list of results objects or a single result object.
Each result object should have a method <cite>plot_qvalues</cite>.</p></li>
<li><p><strong>save_path</strong> (<em>str</em><em>, </em><em>optional</em>) – If provided, saves the plot to the specified path.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – The resolution of the plot. Default is 300.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – The size of the figure. Default is (15, 10).</p></li>
<li><p><strong>threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – The q-value threshold for plotting. Default is 0.05.</p></li>
<li><p><strong>colors</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of colors for the plots. If not provided, uses default colors.</p></li>
<li><p><strong>**kwargs</strong> (<em>dict</em>) – Additional plotting parameters.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function displays or saves the plot.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>This function:
1. Creates a figure with two subplots for PSMs and peptides
2. Plots q-values for each result with different colors
3. Adds legends and titles to each subplot
4. Saves or displays the plot based on save_path</p>
</div>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="usage.html" class="btn btn-neutral float-left" title="Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="examples.html" class="btn btn-neutral float-right" title="Examples" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Zixiang Shang.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>