

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>optimhc.feature_generator.spectra_similarity &mdash; optiMHC 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=01f34227"></script>
      <script src="../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            optiMHC
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">optiMHC</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">optimhc.feature_generator.spectra_similarity</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for optimhc.feature_generator.spectra_similarity</h1><div class="highlight"><pre>
<span></span><span class="c1"># feature_generator/spectra_similarity.py</span>

<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Optional</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">pearsonr</span><span class="p">,</span> <span class="n">spearmanr</span>
<span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">cosine</span>
<span class="kn">from</span> <span class="nn">optimhc.feature_generator.base_feature_generator</span> <span class="kn">import</span> <span class="n">BaseFeatureGenerator</span>
<span class="kn">from</span> <span class="nn">optimhc</span> <span class="kn">import</span> <span class="n">utils</span>
<span class="kn">from</span> <span class="nn">optimhc.parser</span> <span class="kn">import</span> <span class="n">extract_mzml_data</span>
<span class="kn">from</span> <span class="nn">koinapy</span> <span class="kn">import</span> <span class="n">Koina</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator">[docs]</a>
<span class="k">class</span> <span class="nc">SpectraSimilarityFeatureGenerator</span><span class="p">(</span><span class="n">BaseFeatureGenerator</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Feature generator for calculating similarity between experimental and predicted spectra.</span>

<span class="sd">    This class works through the following steps:</span>
<span class="sd">    1. Extract experimental spectral data from mzML files</span>
<span class="sd">    2. Use Koina for theoretical spectra prediction</span>
<span class="sd">    3. Align experimental and predicted spectra</span>
<span class="sd">    4. Calculate similarity metrics as features</span>

<span class="sd">    Parameters:</span>
<span class="sd">        peptides (List[str]): List of peptide sequences</span>
<span class="sd">        charges (List[int]): List of charge states</span>
<span class="sd">        scan_ids (List[int]): List of scan IDs</span>
<span class="sd">        mz_file_paths (List[str]): List of mzML file paths</span>
<span class="sd">        model_type (str): Prediction model type, either &quot;HCD&quot; or &quot;CID&quot;</span>
<span class="sd">        collision_energies (List[float]): List of collision energies, required when model_type is &quot;HCD&quot;</span>
<span class="sd">        remove_pre_nxt_aa (bool): Whether to remove preceding and next amino acids, default is True</span>
<span class="sd">        remove_modification (bool): Whether to remove modifications, default is True</span>
<span class="sd">        url (str): Koina server URL, default is &quot;koina.wilhelmlab.org:443&quot;</span>
<span class="sd">        top_n (int): Number of top peaks to use for alignment, default is 12</span>
<span class="sd">        tolerance_ppm (float): Mass tolerance for alignment in ppm, default is 20</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">spectrum_ids</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">peptides</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">charges</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span>
        <span class="n">scan_ids</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">],</span>
        <span class="n">mz_file_paths</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">model_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">collision_energies</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">instruments</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">fragmentation_types</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">remove_pre_nxt_aa</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">mod_dict</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">url</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;koina.wilhelmlab.org:443&quot;</span><span class="p">,</span>
        <span class="n">top_n</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">36</span><span class="p">,</span>
        <span class="n">tolerance_ppm</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spectrum_ids</span> <span class="o">=</span> <span class="n">spectrum_ids</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">peptides</span> <span class="o">=</span> <span class="n">peptides</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">charges</span> <span class="o">=</span> <span class="n">charges</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">scan_ids</span> <span class="o">=</span> <span class="n">scan_ids</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mz_file_paths</span> <span class="o">=</span> <span class="n">mz_file_paths</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model_type</span> <span class="o">=</span> <span class="n">model_type</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">collision_energies</span> <span class="o">=</span> <span class="n">collision_energies</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">instruments</span> <span class="o">=</span> <span class="n">instruments</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fragmentation_types</span> <span class="o">=</span> <span class="n">fragmentation_types</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">remove_pre_nxt_aa</span> <span class="o">=</span> <span class="n">remove_pre_nxt_aa</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mod_dict</span> <span class="o">=</span> <span class="n">mod_dict</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">url</span> <span class="o">=</span> <span class="n">url</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">top_n</span> <span class="o">=</span> <span class="n">top_n</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tolerance_ppm</span> <span class="o">=</span> <span class="n">tolerance_ppm</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">results</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_raw_predictions</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Initializing SpectraSimilarityFeatureGenerator with </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">peptides</span><span class="p">)</span><span class="si">}</span><span class="s2"> PSMs&quot;</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using model: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">model_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;spectrum_id&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">spectrum_ids</span><span class="p">,</span>
                <span class="s2">&quot;scan&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">scan_ids</span><span class="p">,</span>
                <span class="s2">&quot;peptide&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">peptides</span><span class="p">,</span>
                <span class="s2">&quot;charge&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">charges</span><span class="p">,</span>
                <span class="s2">&quot;mz_file_path&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">mz_file_paths</span><span class="p">,</span>
            <span class="p">}</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;processed_peptide&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;peptide&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_preprocess_peptide</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Recevied </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">)</span><span class="si">}</span><span class="s2"> PSMs for spectral similarity feature generation&quot;</span>
        <span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">id_column</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a list of input columns required for the feature generator.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[</span><span class="s2">&quot;spectrum_id&quot;</span><span class="p">,</span> <span class="s2">&quot;peptide&quot;</span><span class="p">,</span> <span class="s2">&quot;charge&quot;</span><span class="p">]</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">feature_columns</span><span class="p">(</span><span class="n">sself</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a list of feature columns generated by the feature generator.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[</span>
            <span class="s2">&quot;spectral_angle_similarity&quot;</span><span class="p">,</span>
            <span class="s2">&quot;cosine_similarity&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pearson_correlation&quot;</span><span class="p">,</span>
            <span class="s2">&quot;spearman_correlation&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mean_squared_error&quot;</span><span class="p">,</span>
            <span class="s2">&quot;unweighted_entropy_similarity&quot;</span><span class="p">,</span>
            <span class="s2">&quot;predicted_seen_nonzero&quot;</span><span class="p">,</span>
            <span class="s2">&quot;predicted_not_seen&quot;</span><span class="p">,</span>
        <span class="p">]</span>

<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator.input_df">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.input_df">[docs]</a>
    <span class="k">def</span> <span class="nf">input_df</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the generated features as a DataFrame.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: DataFrame containing the generated features</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._preprocess_peptide">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._preprocess_peptide">[docs]</a>
    <span class="k">def</span> <span class="nf">_preprocess_peptide</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">peptide</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Preprocess peptide sequence.</span>

<span class="sd">        As Prosit does not support non-standard amino acid &#39;U&#39;, we replace it with &#39;C&#39;.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        peptide : str</span>
<span class="sd">            Original peptide sequence.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            Processed peptide sequence.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This is nonsense when it comes to spectral prediction, but we need to keep it</span>
<span class="sd">        for compatibility with Koina. In the future, this should be prohibited at the input level.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">processed_peptide</span> <span class="o">=</span> <span class="n">peptide</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_pre_nxt_aa</span><span class="p">:</span>
            <span class="n">processed_peptide</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">remove_pre_and_nxt_aa</span><span class="p">(</span><span class="n">processed_peptide</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mod_dict</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">processed_peptide</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">remove_modifications</span><span class="p">(</span><span class="n">processed_peptide</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">mod</span><span class="p">,</span> <span class="n">replacement</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mod_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">processed_peptide</span> <span class="o">=</span> <span class="n">processed_peptide</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">mod</span><span class="p">,</span> <span class="n">replacement</span><span class="p">)</span>

        <span class="c1"># Replace non-standard amino acid &#39;U&#39; with &#39;C&#39;.</span>
        <span class="c1"># This is nosense when it comes to spectral prediction.</span>
        <span class="c1"># But we need to keep it for compatibility with Koina.</span>
        <span class="c1"># In the future, this should be prohibited at the input level.</span>
        <span class="k">if</span> <span class="s2">&quot;U&quot;</span> <span class="ow">in</span> <span class="n">utils</span><span class="o">.</span><span class="n">remove_modifications</span><span class="p">(</span><span class="n">processed_peptide</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Peptide sequence contains non-standard amino acid &#39;U&#39;: </span><span class="si">{</span><span class="n">processed_peptide</span><span class="si">}</span><span class="s2">. Replacing with &#39;C&#39;.&quot;</span>
            <span class="p">)</span>
            <span class="c1"># Replace &#39;U&#39; with &#39;C&#39;</span>
            <span class="n">processed_peptide</span> <span class="o">=</span> <span class="n">processed_peptide</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;U&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">processed_peptide</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._extract_experimental_spectra">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._extract_experimental_spectra">[docs]</a>
    <span class="k">def</span> <span class="nf">_extract_experimental_spectra</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extract experimental spectral data from mzML files.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            DataFrame containing experimental spectral data.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method groups scan IDs by file path for efficiency and extracts</span>
<span class="sd">        spectral data from each mzML file. The resulting DataFrame contains</span>
<span class="sd">        m/z values, intensities, and associated metadata for each spectrum.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Extracting experimental spectral data...&quot;</span><span class="p">)</span>

        <span class="c1"># Group scan IDs by file path for efficiency</span>
        <span class="n">file_to_scans</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">scan_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;scan&quot;</span><span class="p">]</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;mz_file_path&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">file_path</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">file_to_scans</span><span class="p">:</span>
                <span class="n">file_to_scans</span><span class="p">[</span><span class="n">file_path</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">file_to_scans</span><span class="p">[</span><span class="n">file_path</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">scan_id</span><span class="p">)</span>

        <span class="n">exp_spectra_dfs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">file_path</span><span class="p">,</span> <span class="n">scan_ids</span> <span class="ow">in</span> <span class="n">file_to_scans</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Extracting </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">scan_ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> scans from </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">spectra_df</span> <span class="o">=</span> <span class="n">extract_mzml_data</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">scan_ids</span><span class="p">)</span>
            <span class="n">spectra_df</span><span class="p">[</span><span class="s2">&quot;mz_file_path&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_path</span>
            <span class="n">exp_spectra_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">spectra_df</span><span class="p">)</span>

        <span class="c1"># Merge spectral data from all files</span>
        <span class="k">if</span> <span class="n">exp_spectra_dfs</span><span class="p">:</span>
            <span class="n">exp_spectra_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">exp_spectra_dfs</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Successfully extracted </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">exp_spectra_df</span><span class="p">)</span><span class="si">}</span><span class="s2"> experimental spectra&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="n">exp_spectra_df</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No experimental spectral data found&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._predict_theoretical_spectra">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._predict_theoretical_spectra">[docs]</a>
    <span class="k">def</span> <span class="nf">_predict_theoretical_spectra</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">processed_peptides</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">charges</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Use Koina to predict theoretical spectra.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        processed_peptides : list of str</span>
<span class="sd">            List of preprocessed peptide sequences.</span>
<span class="sd">        charges : list of int</span>
<span class="sd">            List of charge states.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            DataFrame containing predicted spectral data.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        Exception</span>
<span class="sd">            If there is an error during Koina prediction.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method uses Koina to predict theoretical spectra for each peptide.</span>
<span class="sd">        For AlphaPeptDeep_ms2_generic model, inputs are split into batches</span>
<span class="sd">        grouped by peptide length for prediction.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Predicting theoretical spectra using </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">model_type</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">)</span>

        <span class="n">inputs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="n">inputs</span><span class="p">[</span><span class="s2">&quot;peptide_sequences&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">processed_peptides</span><span class="p">)</span>
        <span class="n">inputs</span><span class="p">[</span><span class="s2">&quot;precursor_charges&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">charges</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">collision_energies</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">inputs</span><span class="p">[</span><span class="s2">&quot;collision_energies&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">collision_energies</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">instruments</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">inputs</span><span class="p">[</span><span class="s2">&quot;instrument_types&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">instruments</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">fragmentation_types</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">inputs</span><span class="p">[</span><span class="s2">&quot;fragmentation_types&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fragmentation_types</span><span class="p">)</span>

        <span class="n">model</span> <span class="o">=</span> <span class="n">Koina</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_type</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">url</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">predictions</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">inputs</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error during Koina prediction: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Koina prediction failed. Please check:&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;- Input parameters compatibility&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;- Supported modifications&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;- Peptide length limits&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Details at: https://koina.proteomicsdb.org/&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model_type</span> <span class="o">==</span> <span class="s2">&quot;AlphaPeptDeep_ms2_generic&quot;</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="s2">&quot;Splitting inputs into batches grouped by peptide length for prediction...&quot;</span>
                <span class="p">)</span>
                <span class="kn">import</span> <span class="nn">re</span>

                <span class="c1"># Compute peptide length by removing modifications in square brackets.</span>
                <span class="n">inputs</span><span class="p">[</span><span class="s2">&quot;peptide_length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">inputs</span><span class="p">[</span><span class="s2">&quot;peptide_sequences&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\[.*?\]&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">))</span>
                <span class="p">)</span>

                <span class="n">predictions_batches</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="c1"># Group the DataFrame by calculated peptide length. Each group is submitted as one batch.</span>
                <span class="k">for</span> <span class="n">peptide_length</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">inputs</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;peptide_length&quot;</span><span class="p">):</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Processing </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">group</span><span class="p">)</span><span class="si">}</span><span class="s2"> entries for peptide length </span><span class="si">{</span><span class="n">peptide_length</span><span class="si">}</span><span class="s2">.&quot;</span>
                    <span class="p">)</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">batch_predictions</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">group</span><span class="p">)</span>
                        <span class="n">predictions_batches</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">batch_predictions</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">batch_error</span><span class="p">:</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Error during Koina prediction for batch with peptide length </span><span class="si">{</span><span class="n">peptide_length</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">batch_error</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">)</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Batch data:</span><span class="se">\n</span><span class="si">{</span><span class="n">group</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">raise</span>
                <span class="n">predictions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">predictions_batches</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported model type: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">model_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># Save the raw prediction results</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_raw_predictions</span> <span class="o">=</span> <span class="n">predictions</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

            <span class="c1"># Convert prediction results to a suitable format</span>
            <span class="n">pred_df</span> <span class="o">=</span> <span class="n">predictions</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">pred_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
                <span class="n">columns</span><span class="o">=</span><span class="p">{</span>
                    <span class="s2">&quot;peptide_sequences&quot;</span><span class="p">:</span> <span class="s2">&quot;processed_peptide&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;precursor_charges&quot;</span><span class="p">:</span> <span class="s2">&quot;charge&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;intensities&quot;</span><span class="p">:</span> <span class="s2">&quot;pred_intensity&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;mz&quot;</span><span class="p">:</span> <span class="s2">&quot;pred_mz&quot;</span><span class="p">,</span>
                <span class="p">},</span>
                <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Group by peptide and charge, convert predicted mz and intensity to lists</span>
            <span class="n">grouped_df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">pred_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;processed_peptide&quot;</span><span class="p">,</span> <span class="s2">&quot;charge&quot;</span><span class="p">])</span>
                <span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="s2">&quot;pred_intensity&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="s2">&quot;pred_mz&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="s2">&quot;annotation&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">})</span>
                <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
            <span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Successfully predicted </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">grouped_df</span><span class="p">)</span><span class="si">}</span><span class="s2"> theoretical spectra&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">grouped_df</span></div>


    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">raw_predictions</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the raw prediction results from Koina.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: Raw prediction results DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_raw_predictions</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">results</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generate_features</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_raw_predictions</span>

<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator.get_raw_predictions">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.get_raw_predictions">[docs]</a>
    <span class="k">def</span> <span class="nf">get_raw_predictions</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the raw prediction results DataFrame from Koina.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: Raw prediction results DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">raw_predictions</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator.save_raw_predictions">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.save_raw_predictions">[docs]</a>
    <span class="k">def</span> <span class="nf">save_raw_predictions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Save the raw prediction results to a file.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            file_path (str): Path to save the file</span>
<span class="sd">            **kwargs: Other parameters passed to pandas.DataFrame.to_csv</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="s2">&quot;index&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">raw_predictions</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">raw_predictions</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Raw prediction results saved to: </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No raw prediction results available to save.&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._sort_spectrum_by_mz">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._sort_spectrum_by_mz">[docs]</a>
    <span class="k">def</span> <span class="nf">_sort_spectrum_by_mz</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">mz</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
        <span class="n">intensity</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
        <span class="n">annotation</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">Optional</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sort spectrum (m/z, intensity, and optionally annotation) by m/z values.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        mz : list of float</span>
<span class="sd">            m/z values.</span>
<span class="sd">        intensity : list of float</span>
<span class="sd">            Intensity values.</span>
<span class="sd">        annotation : list of str, optional</span>
<span class="sd">            Fragment annotations.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        tuple of (np.ndarray, np.ndarray, np.ndarray or None)</span>
<span class="sd">            Sorted m/z, intensity, and annotation arrays.</span>
<span class="sd">            If annotation is None, the third element will be None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mz_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mz</span><span class="p">)</span>
        <span class="n">intensity_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">intensity</span><span class="p">)</span>
        <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">mz_array</span><span class="p">)</span>
        <span class="n">sorted_mz</span> <span class="o">=</span> <span class="n">mz_array</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>
        <span class="n">sorted_intensity</span> <span class="o">=</span> <span class="n">intensity_array</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>

        <span class="n">sorted_annotation</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">annotation_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">annotation</span><span class="p">)</span>
            <span class="n">sorted_annotation</span> <span class="o">=</span> <span class="n">annotation_array</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">sorted_mz</span><span class="p">,</span> <span class="n">sorted_intensity</span><span class="p">,</span> <span class="n">sorted_annotation</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._align_spectra_all_peaks">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._align_spectra_all_peaks">[docs]</a>
    <span class="k">def</span> <span class="nf">_align_spectra_all_peaks</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">exp_mz</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
        <span class="n">exp_intensity</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
        <span class="n">pred_mz</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
        <span class="n">pred_intensity</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
        <span class="n">pred_annotation</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">use_ppm</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">],</span> <span class="n">Dict</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Align experimental and predicted spectra.</span>

<span class="sd">        Parameters:</span>
<span class="sd">            exp_mz (List[float]): Experimental m/z values</span>
<span class="sd">            exp_intensity (List[float]): Experimental intensity values</span>
<span class="sd">            pred_mz (List[float]): Predicted m/z values</span>
<span class="sd">            pred_intensity (List[float]): Predicted intensity values</span>
<span class="sd">            pred_annotation (Optional[List[str]]): Predicted fragment annotations</span>
<span class="sd">            use_ppm (bool): Whether to use ppm tolerance or Da tolerance</span>

<span class="sd">        Returns:</span>
<span class="sd">            Tuple[np.ndarray, np.ndarray, List[Tuple], Dict]:</span>
<span class="sd">                - Aligned experimental intensity vector</span>
<span class="sd">                - Predicted intensity vector</span>
<span class="sd">                - Matching index pairs</span>
<span class="sd">                - Additional info including original sorted arrays</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Sort both experimental and predicted spectra by m/z</span>
        <span class="n">exp_mz_sorted</span><span class="p">,</span> <span class="n">exp_intensity_sorted</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sort_spectrum_by_mz</span><span class="p">(</span>
            <span class="n">exp_mz</span><span class="p">,</span> <span class="n">exp_intensity</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">pred_annotation</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">pred_mz_sorted</span><span class="p">,</span> <span class="n">pred_intensity_sorted</span><span class="p">,</span> <span class="n">pred_annotation_sorted</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_sort_spectrum_by_mz</span><span class="p">(</span><span class="n">pred_mz</span><span class="p">,</span> <span class="n">pred_intensity</span><span class="p">,</span> <span class="n">pred_annotation</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">pred_mz_sorted</span><span class="p">,</span> <span class="n">pred_intensity_sorted</span><span class="p">,</span> <span class="n">pred_annotation_sorted</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_sort_spectrum_by_mz</span><span class="p">(</span><span class="n">pred_mz</span><span class="p">,</span> <span class="n">pred_intensity</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="n">aligned_exp_intensity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pred_mz_sorted</span><span class="p">))</span>
        <span class="n">aligned_pred_intensity</span> <span class="o">=</span> <span class="n">pred_intensity_sorted</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">matched_indices</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">start_pos</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">pred_peak_mz</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pred_mz_sorted</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">use_ppm</span><span class="p">:</span>
                <span class="n">fragment_min</span> <span class="o">=</span> <span class="n">pred_peak_mz</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">tolerance_ppm</span> <span class="o">/</span> <span class="mf">1e6</span><span class="p">)</span>
                <span class="n">fragment_max</span> <span class="o">=</span> <span class="n">pred_peak_mz</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">tolerance_ppm</span> <span class="o">/</span> <span class="mf">1e6</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># If using Da tolerance, use a fixed window (typically ~0.05 Da)</span>
                <span class="n">tolerance</span> <span class="o">=</span> <span class="mf">0.05</span>  <span class="c1"># Default value</span>
                <span class="n">fragment_min</span> <span class="o">=</span> <span class="n">pred_peak_mz</span> <span class="o">-</span> <span class="n">tolerance</span>
                <span class="n">fragment_max</span> <span class="o">=</span> <span class="n">pred_peak_mz</span> <span class="o">+</span> <span class="n">tolerance</span>

            <span class="n">matched_int</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">matched_exp_idx</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">past_start</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="k">while</span> <span class="n">start_pos</span> <span class="o">+</span> <span class="n">past_start</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">exp_mz_sorted</span><span class="p">):</span>
                <span class="n">exp_peak_mz</span> <span class="o">=</span> <span class="n">exp_mz_sorted</span><span class="p">[</span><span class="n">start_pos</span> <span class="o">+</span> <span class="n">past_start</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">exp_peak_mz</span> <span class="o">&lt;</span> <span class="n">fragment_min</span><span class="p">:</span>
                    <span class="n">start_pos</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">exp_peak_mz</span> <span class="o">&lt;=</span> <span class="n">fragment_max</span><span class="p">:</span>
                    <span class="n">exp_peak_int</span> <span class="o">=</span> <span class="n">exp_intensity_sorted</span><span class="p">[</span><span class="n">start_pos</span> <span class="o">+</span> <span class="n">past_start</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">exp_peak_int</span> <span class="o">&gt;</span> <span class="n">matched_int</span><span class="p">:</span>
                        <span class="n">matched_int</span> <span class="o">=</span> <span class="n">exp_peak_int</span>
                        <span class="n">matched_exp_idx</span> <span class="o">=</span> <span class="n">start_pos</span> <span class="o">+</span> <span class="n">past_start</span>
                    <span class="n">past_start</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">break</span>

            <span class="n">aligned_exp_intensity</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">matched_int</span>

            <span class="c1"># Record matching index pairs (pred_idx, exp_idx)</span>
            <span class="n">pred_idx</span> <span class="o">=</span> <span class="n">i</span>
            <span class="n">exp_idx</span> <span class="o">=</span> <span class="n">matched_exp_idx</span>
            <span class="n">matched_indices</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">pred_idx</span><span class="p">,</span> <span class="n">exp_idx</span><span class="p">))</span>

        <span class="n">additional_info</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;exp_mz_sorted&quot;</span><span class="p">:</span> <span class="n">exp_mz_sorted</span><span class="p">,</span>
            <span class="s2">&quot;exp_intensity_sorted&quot;</span><span class="p">:</span> <span class="n">exp_intensity_sorted</span><span class="p">,</span>
            <span class="s2">&quot;pred_mz_sorted&quot;</span><span class="p">:</span> <span class="n">pred_mz_sorted</span><span class="p">,</span>
            <span class="s2">&quot;pred_intensity_sorted&quot;</span><span class="p">:</span> <span class="n">pred_intensity_sorted</span><span class="p">,</span>
            <span class="s2">&quot;pred_annotation_sorted&quot;</span><span class="p">:</span> <span class="n">pred_annotation_sorted</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="n">aligned_exp_intensity</span><span class="p">,</span>
            <span class="n">aligned_pred_intensity</span><span class="p">,</span>
            <span class="n">matched_indices</span><span class="p">,</span>
            <span class="n">additional_info</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._get_top_peaks_vectors">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._get_top_peaks_vectors">[docs]</a>
    <span class="k">def</span> <span class="nf">_get_top_peaks_vectors</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">aligned_exp_intensity</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">aligned_pred_intensity</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">matched_indices</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">],</span>
        <span class="n">top_n</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extract top N peaks based on predicted intensity for similarity calculation</span>

<span class="sd">        Parameters:</span>
<span class="sd">            aligned_exp_intensity (np.ndarray): Aligned experimental intensity vector</span>
<span class="sd">            aligned_pred_intensity (np.ndarray): Aligned predicted intensity vector</span>
<span class="sd">            matched_indices (List[Tuple]): Matching index pairs (pred_idx, exp_idx)</span>
<span class="sd">            top_n (int): Number of top peaks to extract</span>

<span class="sd">        Returns:</span>
<span class="sd">            Tuple[np.ndarray, np.ndarray, List[Tuple]]:</span>
<span class="sd">                - Top N experimental intensity vector</span>
<span class="sd">                - Top N predicted intensity vector</span>
<span class="sd">                - Top N matching index pairs</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Get indices of top N peaks by predicted intensity</span>
        <span class="n">num_peaks</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">top_n</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">aligned_pred_intensity</span><span class="p">))</span>
        <span class="n">top_pred_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">aligned_pred_intensity</span><span class="p">)[:</span><span class="n">num_peaks</span><span class="p">]</span>

        <span class="c1"># Extract top N peaks</span>
        <span class="n">top_exp_intensity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">num_peaks</span><span class="p">)</span>
        <span class="n">top_pred_intensity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">num_peaks</span><span class="p">)</span>
        <span class="n">top_matched_indices</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">orig_idx</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">top_pred_indices</span><span class="p">):</span>
            <span class="n">top_exp_intensity</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">aligned_exp_intensity</span><span class="p">[</span><span class="n">orig_idx</span><span class="p">]</span>
            <span class="n">top_pred_intensity</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">aligned_pred_intensity</span><span class="p">[</span><span class="n">orig_idx</span><span class="p">]</span>
            <span class="n">top_matched_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">matched_indices</span><span class="p">[</span><span class="n">orig_idx</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">top_exp_intensity</span><span class="p">,</span> <span class="n">top_pred_intensity</span><span class="p">,</span> <span class="n">top_matched_indices</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._align_spectra">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._align_spectra">[docs]</a>
    <span class="k">def</span> <span class="nf">_align_spectra</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">exp_mz</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
        <span class="n">exp_intensity</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
        <span class="n">pred_mz</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
        <span class="n">pred_intensity</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
        <span class="n">pred_annotation</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Align experimental and predicted spectra.</span>

<span class="sd">        This is a wrapper around _align_spectra_all_peaks and _get_top_peaks_vectors</span>
<span class="sd">        to maintain backward compatibility while using the improved algorithm.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        exp_mz : list of float</span>
<span class="sd">            Experimental m/z values.</span>
<span class="sd">        exp_intensity : list of float</span>
<span class="sd">            Experimental intensity values.</span>
<span class="sd">        pred_mz : list of float</span>
<span class="sd">            Predicted m/z values.</span>
<span class="sd">        pred_intensity : list of float</span>
<span class="sd">            Predicted intensity values.</span>
<span class="sd">        pred_annotation : list of str, optional</span>
<span class="sd">            Predicted fragment annotations.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        tuple of (np.ndarray, np.ndarray, list of tuple)</span>
<span class="sd">            - Aligned experimental intensity vector</span>
<span class="sd">            - Predicted intensity vector</span>
<span class="sd">            - Matching index pairs (for top N peaks)</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The method first aligns all peaks using _align_spectra_all_peaks, then</span>
<span class="sd">        extracts the top N peaks using _get_top_peaks_vectors for compatibility</span>
<span class="sd">        with existing code.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># First align all peaks</span>
        <span class="n">all_exp_intensity</span><span class="p">,</span> <span class="n">all_pred_intensity</span><span class="p">,</span> <span class="n">all_matched_indices</span><span class="p">,</span> <span class="n">additional_info</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_align_spectra_all_peaks</span><span class="p">(</span>
                <span class="n">exp_mz</span><span class="p">,</span>
                <span class="n">exp_intensity</span><span class="p">,</span>
                <span class="n">pred_mz</span><span class="p">,</span>
                <span class="n">pred_intensity</span><span class="p">,</span>
                <span class="n">pred_annotation</span><span class="p">,</span>
                <span class="n">use_ppm</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Then get top N peaks for compatibility with existing code</span>
        <span class="n">top_exp_intensity</span><span class="p">,</span> <span class="n">top_pred_intensity</span><span class="p">,</span> <span class="n">top_matched_indices</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_get_top_peaks_vectors</span><span class="p">(</span>
                <span class="n">all_exp_intensity</span><span class="p">,</span> <span class="n">all_pred_intensity</span><span class="p">,</span> <span class="n">all_matched_indices</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">top_n</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">top_exp_intensity</span><span class="p">,</span> <span class="n">top_pred_intensity</span><span class="p">,</span> <span class="n">top_matched_indices</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._normalize_vector_l2">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._normalize_vector_l2">[docs]</a>
    <span class="k">def</span> <span class="nf">_normalize_vector_l2</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Normalize a vector using L2 normalization (unit vector).</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vector : np.ndarray</span>
<span class="sd">            Input vector to normalize.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        np.ndarray</span>
<span class="sd">            L2-normalized vector.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        If the input vector has zero norm, the original vector is returned unchanged.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">norm</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">vector</span><span class="p">)</span>  <span class="c1"># Calculate the L2 norm (Euclidean norm)</span>
        <span class="k">if</span> <span class="n">norm</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">vector</span>
        <span class="k">return</span> <span class="n">vector</span> <span class="o">/</span> <span class="n">norm</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._normalize_vector_sum">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._normalize_vector_sum">[docs]</a>
    <span class="k">def</span> <span class="nf">_normalize_vector_sum</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Perform sum normalization (probability normalization) on a vector.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vector : np.ndarray</span>
<span class="sd">            Input vector.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        np.ndarray</span>
<span class="sd">            Sum normalized vector (sum to 1).</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        If the input vector has zero sum, the original vector is returned unchanged.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">total</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">vector</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">total</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">vector</span> <span class="o">/</span> <span class="n">total</span>
        <span class="k">return</span> <span class="n">vector</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._calculate_spectral_angle_similarity">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_spectral_angle_similarity">[docs]</a>
    <span class="k">def</span> <span class="nf">_calculate_spectral_angle_similarity</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exp_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the spectral angle between experimental and predicted vectors.</span>

<span class="sd">        Normalize the angle to [0, 1] where 1 is the best similarity.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        exp_vector : np.ndarray</span>
<span class="sd">            Experimental intensity vector.</span>
<span class="sd">        pred_vector : np.ndarray</span>
<span class="sd">            Predicted intensity vector.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Spectral angle similarity (0-1, higher is better).</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The spectral angle is calculated as: SA = 1 - (2 * angle / ),</span>
<span class="sd">        where angle is the angle between the normalized vectors in radians.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">exp_norm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_vector_l2</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">)</span>
        <span class="n">pred_norm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_vector_l2</span><span class="p">(</span><span class="n">pred_vector</span><span class="p">)</span>
        <span class="n">dot_product</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">exp_norm</span> <span class="o">*</span> <span class="n">pred_norm</span><span class="p">)</span>

        <span class="c1"># Clamp dot product to [-1, 1] to avoid numerical issues</span>
        <span class="n">dot_product</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">dot_product</span><span class="p">,</span> <span class="o">-</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
        <span class="n">angle</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arccos</span><span class="p">(</span><span class="n">dot_product</span><span class="p">)</span>  <span class="c1"># Angle in radians</span>

        <span class="c1"># Return similarity score: SA = 1 - (2 * angle / )</span>
        <span class="k">return</span> <span class="mi">1</span> <span class="o">-</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">angle</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">pi</span><span class="p">)</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._calculate_cosine_similarity">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_cosine_similarity">[docs]</a>
    <span class="k">def</span> <span class="nf">_calculate_cosine_similarity</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exp_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate the cosine similarity between experimental and predicted vectors.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        exp_vector : np.ndarray</span>
<span class="sd">            Experimental intensity vector.</span>
<span class="sd">        pred_vector : np.ndarray</span>
<span class="sd">            Predicted intensity vector.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Cosine similarity (0-1, higher is better).</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The cosine similarity is calculated as 1 - cosine_distance.</span>
<span class="sd">        If either vector has zero sum, returns 0.0.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">pred_vector</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="mf">0.0</span>

        <span class="c1"># Normalize vectors using L2 normalization</span>
        <span class="n">exp_norm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_vector_l2</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">)</span>
        <span class="n">pred_norm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_vector_l2</span><span class="p">(</span><span class="n">pred_vector</span><span class="p">)</span>

        <span class="c1"># Use 1 - cosine distance = cosine similarity</span>
        <span class="k">return</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">cosine</span><span class="p">(</span><span class="n">exp_norm</span><span class="p">,</span> <span class="n">pred_norm</span><span class="p">)</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._calculate_spearman_correlation">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_spearman_correlation">[docs]</a>
    <span class="k">def</span> <span class="nf">_calculate_spearman_correlation</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exp_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate Spearman correlation coefficient between experimental and predicted vectors.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        exp_vector : np.ndarray</span>
<span class="sd">            Experimental intensity vector.</span>
<span class="sd">        pred_vector : np.ndarray</span>
<span class="sd">            Predicted intensity vector.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Spearman correlation coefficient (-1 to 1, higher is better).</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        If either vector has no variation (std = 0), returns 0.0.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Handle vectors with no variation</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">pred_vector</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="mf">0.0</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">r</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">spearmanr</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">r</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">return</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._calculate_pearson_correlation">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_pearson_correlation">[docs]</a>
    <span class="k">def</span> <span class="nf">_calculate_pearson_correlation</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exp_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate Pearson correlation coefficient between experimental and predicted vectors.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        exp_vector : np.ndarray</span>
<span class="sd">            Experimental intensity vector.</span>
<span class="sd">        pred_vector : np.ndarray</span>
<span class="sd">            Predicted intensity vector.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Pearson correlation coefficient (-1 to 1, higher is better).</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        If either vector has no variation (std = 0), returns 0.0.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Handle vectors with no variation</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">pred_vector</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="mf">0.0</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">r</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">pearsonr</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">r</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">return</span> <span class="mf">0.0</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._calculate_mean_squared_error">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_mean_squared_error">[docs]</a>
    <span class="k">def</span> <span class="nf">_calculate_mean_squared_error</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exp_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate mean squared error between experimental and predicted vectors.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        exp_vector : np.ndarray</span>
<span class="sd">            Experimental intensity vector.</span>
<span class="sd">        pred_vector : np.ndarray</span>
<span class="sd">            Predicted intensity vector.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Mean squared error (lower is better).</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The vectors are normalized using L2 normalization before calculating</span>
<span class="sd">        the mean squared error for fair comparison.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Normalize vectors for fair comparison using L2 normalization</span>
        <span class="n">exp_norm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_vector_l2</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">)</span>
        <span class="n">pred_norm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_vector_l2</span><span class="p">(</span><span class="n">pred_vector</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">((</span><span class="n">exp_norm</span> <span class="o">-</span> <span class="n">pred_norm</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._calculate_entropy">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_entropy">[docs]</a>
    <span class="k">def</span> <span class="nf">_calculate_entropy</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate Shannon entropy of a vector that has already been sum-1 normalized.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vector : np.ndarray</span>
<span class="sd">            Input vector, which has already been sum-1 normalized.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Shannon entropy.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Only non-zero probabilities are considered for entropy calculation.</span>
<span class="sd">        If all probabilities are zero, returns 0.0.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Only consider non-zero probabilities for entropy calculation</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">vector</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">mask</span><span class="p">):</span>
            <span class="k">return</span> <span class="mf">0.0</span>

        <span class="n">prob_vector</span> <span class="o">=</span> <span class="n">vector</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
        <span class="k">return</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">prob_vector</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">prob_vector</span><span class="p">))</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._calculate_unweighted_entropy_similarity">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_unweighted_entropy_similarity">[docs]</a>
    <span class="k">def</span> <span class="nf">_calculate_unweighted_entropy_similarity</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exp_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate unweighted spectral entropy between experimental and predicted vectors.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        exp_vector : np.ndarray</span>
<span class="sd">            Experimental intensity vector.</span>
<span class="sd">        pred_vector : np.ndarray</span>
<span class="sd">            Predicted intensity vector.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        float</span>
<span class="sd">            Spectral entropy similarity.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Based on the method described in https://www.nature.com/articles/s41592-021-01331-z.</span>
<span class="sd">        The spectral entropy is calculated using the formula:</span>
<span class="sd">        1 - (2*S_PM - S_P - S_M)/ln(4), where S_PM is the entropy of the mixed</span>
<span class="sd">        distribution, and S_P and S_M are the entropies of the individual distributions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Sum-to-1 normalization</span>
        <span class="n">sum_normalized_exp_vector</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_vector_sum</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">)</span>
        <span class="n">sum_normalized_pred_vector</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_vector_sum</span><span class="p">(</span><span class="n">pred_vector</span><span class="p">)</span>
        <span class="n">s_exp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_entropy</span><span class="p">(</span><span class="n">sum_normalized_exp_vector</span><span class="p">)</span>
        <span class="n">s_pred</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_entropy</span><span class="p">(</span><span class="n">sum_normalized_pred_vector</span><span class="p">)</span>
        <span class="n">s_mixed</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_entropy</span><span class="p">(</span>
            <span class="mf">0.5</span> <span class="o">*</span> <span class="p">(</span><span class="n">sum_normalized_exp_vector</span> <span class="o">+</span> <span class="n">sum_normalized_pred_vector</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Calculate spectral entropy using formula: 1 - (2*S_PM - S_P - S_M)/ln(4)</span>
        <span class="n">unweighted_entropy_similarity</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">s_mixed</span> <span class="o">-</span> <span class="n">s_exp</span> <span class="o">-</span> <span class="n">s_pred</span><span class="p">)</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">unweighted_entropy_similarity</span></div>


    <span class="c1"># TODO: Assign weight to each peak based on the entropy of a spectrum.</span>
    <span class="c1"># The original paper uses 3 as a entropy cutoff to assign more weight to the low intensity peaks.</span>
    <span class="c1"># But the cutoff value is determined by a small-molecular dataset rather than a proteomics dataset.</span>
    <span class="c1"># Thus, we need to design our own heuristic algorithm to calculate a similar feature for proteomics dataset.</span>
    <span class="c1"># We can refer to the practice of MSBooster.</span>
    <span class="c1"># url: https://github.com/Nesvilab/MSBooster/blame/master/src/main/java/features/spectra/SpectrumComparison.java#L803</span>

    <span class="c1"># def _calculate_entropy_similarity(self, exp_vector: np.ndarray, pred_vector: np.ndarray) -&gt; float:</span>

<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._calculate_predicted_counts">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_predicted_counts">[docs]</a>
    <span class="k">def</span> <span class="nf">_calculate_predicted_counts</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exp_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate counts of predicted peaks seen/not seen in experimental spectrum.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        exp_vector : np.ndarray</span>
<span class="sd">            Experimental intensity vector.</span>
<span class="sd">        pred_vector : np.ndarray</span>
<span class="sd">            Predicted intensity vector.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        tuple of (int, int)</span>
<span class="sd">            - predicted_seen_nonzero: Number of predicted peaks that are also present in experimental spectrum</span>
<span class="sd">            - predicted_not_seen: Number of predicted peaks that are not present in experimental spectrum</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">predicted_seen_nonzero</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">pred_vector</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">exp_vector</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">))</span>
        <span class="n">predicted_not_seen</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">pred_vector</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">exp_vector</span> <span class="o">==</span> <span class="mi">0</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">predicted_seen_nonzero</span><span class="p">,</span> <span class="n">predicted_not_seen</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._calculate_bray_curtis_similarity">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_bray_curtis_similarity">[docs]</a>
    <span class="k">def</span> <span class="nf">_calculate_bray_curtis_similarity</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exp_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate Bray-Curtis similarity between experimental and predicted vectors</span>

<span class="sd">        Parameters:</span>
<span class="sd">            exp_vector (np.ndarray): Experimental intensity vector</span>
<span class="sd">            pred_vector (np.ndarray): Predicted intensity vector</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Bray-Curtis similarity (0-1, higher is better)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Bray-Curtis dissimilarity: sum(|exp - pred|) / sum(exp + pred)</span>
        <span class="n">numerator</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">exp_vector</span> <span class="o">-</span> <span class="n">pred_vector</span><span class="p">))</span>
        <span class="n">denominator</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">exp_vector</span> <span class="o">+</span> <span class="n">pred_vector</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">denominator</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="mf">0.0</span>

        <span class="c1"># Convert dissimilarity to similarity: 1 - dissimilarity</span>
        <span class="k">return</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="p">(</span><span class="n">numerator</span> <span class="o">/</span> <span class="n">denominator</span><span class="p">)</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._calculate_similarity_features">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._calculate_similarity_features">[docs]</a>
    <span class="k">def</span> <span class="nf">_calculate_similarity_features</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">exp_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate all similarity features between experimental and predicted vectors.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        exp_vector : np.ndarray</span>
<span class="sd">            Experimental intensity vector.</span>
<span class="sd">        pred_vector : np.ndarray</span>
<span class="sd">            Predicted intensity vector.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        dict of str to float</span>
<span class="sd">            Dictionary of similarity features, including:</span>
<span class="sd">            - spectral_angle_similarity: Spectral angle similarity (0-1)</span>
<span class="sd">            - cosine_similarity: Cosine similarity (0-1)</span>
<span class="sd">            - pearson_correlation: Pearson correlation (-1 to 1)</span>
<span class="sd">            - spearman_correlation: Spearman correlation (-1 to 1)</span>
<span class="sd">            - mean_squared_error: Mean squared error</span>
<span class="sd">            - unweighted_entropy_similarity: Spectral entropy similarity</span>
<span class="sd">            - predicted_seen_nonzero: Number of predicted peaks seen in experimental spectrum</span>
<span class="sd">            - predicted_not_seen: Number of predicted peaks not seen in experimental spectrum</span>
<span class="sd">            - bray_curtis_similarity: Bray-Curtis similarity (0-1)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">spectral_angle_similarity</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_spectral_angle_similarity</span><span class="p">(</span>
            <span class="n">exp_vector</span><span class="p">,</span> <span class="n">pred_vector</span>
        <span class="p">)</span>
        <span class="n">cosine_similarity</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_cosine_similarity</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">)</span>
        <span class="n">pearson_correlation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_pearson_correlation</span><span class="p">(</span>
            <span class="n">exp_vector</span><span class="p">,</span> <span class="n">pred_vector</span>
        <span class="p">)</span>
        <span class="n">spearman_correlation</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_spearman_correlation</span><span class="p">(</span>
            <span class="n">exp_vector</span><span class="p">,</span> <span class="n">pred_vector</span>
        <span class="p">)</span>
        <span class="n">mean_squared_error</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_mean_squared_error</span><span class="p">(</span><span class="n">exp_vector</span><span class="p">,</span> <span class="n">pred_vector</span><span class="p">)</span>
        <span class="n">unweighted_entropy_similarity</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_unweighted_entropy_similarity</span><span class="p">(</span>
            <span class="n">exp_vector</span><span class="p">,</span> <span class="n">pred_vector</span>
        <span class="p">)</span>
        <span class="n">predicted_seen_nonzero</span><span class="p">,</span> <span class="n">predicted_not_seen</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_predicted_counts</span><span class="p">(</span>
            <span class="n">exp_vector</span><span class="p">,</span> <span class="n">pred_vector</span>
        <span class="p">)</span>
        <span class="n">bray_curtis_similarity</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_bray_curtis_similarity</span><span class="p">(</span>
            <span class="n">exp_vector</span><span class="p">,</span> <span class="n">pred_vector</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span>
            <span class="s2">&quot;spectral_angle_similarity&quot;</span><span class="p">:</span> <span class="n">spectral_angle_similarity</span><span class="p">,</span>
            <span class="s2">&quot;cosine_similarity&quot;</span><span class="p">:</span> <span class="n">cosine_similarity</span><span class="p">,</span>
            <span class="s2">&quot;pearson_correlation&quot;</span><span class="p">:</span> <span class="n">pearson_correlation</span><span class="p">,</span>
            <span class="s2">&quot;spearman_correlation&quot;</span><span class="p">:</span> <span class="n">spearman_correlation</span><span class="p">,</span>
            <span class="s2">&quot;mean_squared_error&quot;</span><span class="p">:</span> <span class="n">mean_squared_error</span><span class="p">,</span>
            <span class="s2">&quot;unweighted_entropy_similarity&quot;</span><span class="p">:</span> <span class="n">unweighted_entropy_similarity</span><span class="p">,</span>
            <span class="s2">&quot;predicted_seen_nonzero&quot;</span><span class="p">:</span> <span class="n">predicted_seen_nonzero</span><span class="p">,</span>
            <span class="s2">&quot;predicted_not_seen&quot;</span><span class="p">:</span> <span class="n">predicted_not_seen</span><span class="p">,</span>
            <span class="s2">&quot;bray_curtis_similarity&quot;</span><span class="p">:</span> <span class="n">bray_curtis_similarity</span><span class="p">,</span>
        <span class="p">}</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator._generate_features">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator._generate_features">[docs]</a>
    <span class="k">def</span> <span class="nf">_generate_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate spectral similarity features</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: DataFrame containing generated features</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">psm_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">pred_spectra_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_predict_theoretical_spectra</span><span class="p">(</span>
            <span class="n">processed_peptides</span><span class="o">=</span><span class="n">psm_df</span><span class="p">[</span><span class="s2">&quot;processed_peptide&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
            <span class="n">charges</span><span class="o">=</span><span class="n">psm_df</span><span class="p">[</span><span class="s2">&quot;charge&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
        <span class="p">)</span>

        <span class="n">exp_spectra_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_extract_experimental_spectra</span><span class="p">()</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">exp_spectra_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">psm_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
                <span class="n">psm_df</span><span class="p">,</span>
                <span class="n">exp_spectra_df</span><span class="p">,</span>
                <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;scan&quot;</span><span class="p">,</span> <span class="s2">&quot;mz_file_path&quot;</span><span class="p">,</span> <span class="s2">&quot;charge&quot;</span><span class="p">],</span>
                <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
                <span class="n">validate</span><span class="o">=</span><span class="s2">&quot;m:1&quot;</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="s2">&quot;Could not extract experimental spectral data, cannot continue processing&quot;</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">psm_df</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Some PSMs were not found in experimental spectral data&quot;</span><span class="p">)</span>

        <span class="n">psm_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">psm_df</span><span class="p">,</span> <span class="n">pred_spectra_df</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;processed_peptide&quot;</span><span class="p">,</span> <span class="s2">&quot;charge&quot;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span>
        <span class="p">)</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Matching experimental and predicted spectra... This may take a while.&quot;</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">psm_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">exp_mz</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;mz&quot;</span><span class="p">]</span>
            <span class="n">exp_intensity</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;intensity&quot;</span><span class="p">]</span>
            <span class="n">pred_mz</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;pred_mz&quot;</span><span class="p">]</span>
            <span class="n">pred_intensity</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;pred_intensity&quot;</span><span class="p">]</span>
            <span class="n">pred_annotation</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;annotation&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="s2">&quot;annotation&quot;</span> <span class="ow">in</span> <span class="n">row</span> <span class="k">else</span> <span class="kc">None</span>

            <span class="c1"># Align all peaks</span>
            <span class="p">(</span>
                <span class="n">all_exp_intensity</span><span class="p">,</span>
                <span class="n">all_pred_intensity</span><span class="p">,</span>
                <span class="n">all_matched_indices</span><span class="p">,</span>
                <span class="n">additional_info</span><span class="p">,</span>
            <span class="p">)</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_align_spectra_all_peaks</span><span class="p">(</span>
                <span class="n">exp_mz</span><span class="p">,</span>
                <span class="n">exp_intensity</span><span class="p">,</span>
                <span class="n">pred_mz</span><span class="p">,</span>
                <span class="n">pred_intensity</span><span class="p">,</span>
                <span class="n">pred_annotation</span><span class="p">,</span>
                <span class="n">use_ppm</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Extract top N peaks for similarity calculations</span>
            <span class="n">top_exp_intensity</span><span class="p">,</span> <span class="n">top_pred_intensity</span><span class="p">,</span> <span class="n">top_matched_indices</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_get_top_peaks_vectors</span><span class="p">(</span>
                    <span class="n">all_exp_intensity</span><span class="p">,</span>
                    <span class="n">all_pred_intensity</span><span class="p">,</span>
                    <span class="n">all_matched_indices</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">top_n</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>

            <span class="n">similarity_features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_similarity_features</span><span class="p">(</span>
                <span class="n">top_exp_intensity</span><span class="p">,</span> <span class="n">top_pred_intensity</span>
            <span class="p">)</span>
            <span class="n">result</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;exp_vector&quot;</span><span class="p">:</span> <span class="n">all_exp_intensity</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                <span class="s2">&quot;pred_vector&quot;</span><span class="p">:</span> <span class="n">all_pred_intensity</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                <span class="s2">&quot;matched_indices&quot;</span><span class="p">:</span> <span class="n">all_matched_indices</span><span class="p">,</span>
                <span class="s2">&quot;exp_top_vector&quot;</span><span class="p">:</span> <span class="n">top_exp_intensity</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                <span class="s2">&quot;pred_top_vector&quot;</span><span class="p">:</span> <span class="n">top_pred_intensity</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                <span class="s2">&quot;top_matched_indices&quot;</span><span class="p">:</span> <span class="n">top_matched_indices</span><span class="p">,</span>
                <span class="o">**</span><span class="n">similarity_features</span><span class="p">,</span>
                <span class="s2">&quot;exp_mz_sorted&quot;</span><span class="p">:</span> <span class="n">additional_info</span><span class="p">[</span><span class="s2">&quot;exp_mz_sorted&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                <span class="s2">&quot;exp_intensity_sorted&quot;</span><span class="p">:</span> <span class="n">additional_info</span><span class="p">[</span>
                    <span class="s2">&quot;exp_intensity_sorted&quot;</span>
                <span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                <span class="s2">&quot;pred_mz_sorted&quot;</span><span class="p">:</span> <span class="n">additional_info</span><span class="p">[</span><span class="s2">&quot;pred_mz_sorted&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                <span class="s2">&quot;pred_intensity_sorted&quot;</span><span class="p">:</span> <span class="n">additional_info</span><span class="p">[</span>
                    <span class="s2">&quot;pred_intensity_sorted&quot;</span>
                <span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
            <span class="p">}</span>

            <span class="c1"># Add annotations if available</span>
            <span class="k">if</span> <span class="n">additional_info</span><span class="p">[</span><span class="s2">&quot;pred_annotation_sorted&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">result</span><span class="p">[</span><span class="s2">&quot;pred_annotation_sorted&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">additional_info</span><span class="p">[</span>
                    <span class="s2">&quot;pred_annotation_sorted&quot;</span>
                <span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

            <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

        <span class="n">results_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
        <span class="n">psm_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">[</span><span class="n">psm_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="n">results_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
        <span class="p">)</span>

        <span class="n">result_columns</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;mz_file_path&quot;</span><span class="p">,</span>
            <span class="s2">&quot;spectrum_id&quot;</span><span class="p">,</span>
            <span class="s2">&quot;scan&quot;</span><span class="p">,</span>
            <span class="s2">&quot;peptide&quot;</span><span class="p">,</span>
            <span class="s2">&quot;charge&quot;</span><span class="p">,</span>
            <span class="s2">&quot;processed_peptide&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mz&quot;</span><span class="p">,</span>
            <span class="s2">&quot;intensity&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pred_mz&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pred_intensity&quot;</span><span class="p">,</span>
            <span class="s2">&quot;annotation&quot;</span><span class="p">,</span>
            <span class="s2">&quot;exp_mz_sorted&quot;</span><span class="p">,</span>
            <span class="s2">&quot;exp_intensity_sorted&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pred_mz_sorted&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pred_intensity_sorted&quot;</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="s2">&quot;pred_annotation_sorted&quot;</span>
                <span class="k">if</span> <span class="s2">&quot;pred_annotation_sorted&quot;</span> <span class="ow">in</span> <span class="n">psm_df</span><span class="o">.</span><span class="n">columns</span>
                <span class="k">else</span> <span class="kc">None</span>
            <span class="p">),</span>
            <span class="s2">&quot;exp_vector&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pred_vector&quot;</span><span class="p">,</span>
            <span class="s2">&quot;matched_indices&quot;</span><span class="p">,</span>
            <span class="s2">&quot;exp_top_vector&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pred_top_vector&quot;</span><span class="p">,</span>
            <span class="s2">&quot;top_matched_indices&quot;</span><span class="p">,</span>
            <span class="s2">&quot;spectral_angle_similarity&quot;</span><span class="p">,</span>
            <span class="s2">&quot;cosine_similarity&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pearson_correlation&quot;</span><span class="p">,</span>
            <span class="s2">&quot;spearman_correlation&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mean_squared_error&quot;</span><span class="p">,</span>
            <span class="s2">&quot;unweighted_entropy_similarity&quot;</span><span class="p">,</span>
            <span class="s2">&quot;predicted_seen_nonzero&quot;</span><span class="p">,</span>
            <span class="s2">&quot;predicted_not_seen&quot;</span><span class="p">,</span>
        <span class="p">]</span>

        <span class="n">result_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">result_columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>
        <span class="n">result_df</span> <span class="o">=</span> <span class="n">psm_df</span><span class="p">[</span><span class="n">result_columns</span><span class="p">]</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Successfully generated spectral similarity features for </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">result_df</span><span class="p">)</span><span class="si">}</span><span class="s2"> PSMs&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">result_df</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator.generate_features">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.generate_features">[docs]</a>
    <span class="k">def</span> <span class="nf">generate_features</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Public interface for generating spectral similarity features.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            DataFrame containing the generated features.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This method is a wrapper around _generate_features that ensures</span>
<span class="sd">        the results are cached and only computed once.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">results</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generate_features</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">id_column</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_columns</span><span class="p">]</span></div>


<div class="viewcode-block" id="SpectraSimilarityFeatureGenerator.get_full_data">
<a class="viewcode-back" href="../../../api.html#optimhc.feature_generator.spectra_similarity.SpectraSimilarityFeatureGenerator.get_full_data">[docs]</a>
    <span class="k">def</span> <span class="nf">get_full_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the full DataFrame with all columns.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pd.DataFrame</span>
<span class="sd">            Full DataFrame with all columns.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This method returns the complete DataFrame including all intermediate</span>
<span class="sd">        results and raw data used in feature generation.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">results</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Zixiang Shang.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>